{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bfa9a3da-a6fa-4ee9-979c-acc7d50cbabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "from corda import reaction_confidence, CORDA\n",
    "from cobra.io import load_matlab_model, save_matlab_model\n",
    "mod_PRrod = load_matlab_model(\"C:/Users/prins/git/Human1_RPE_PR/cs_mods/mod_PRrod.mat\")\n",
    "mod_RPE = load_matlab_model(\"C:/Users/prins/git/Human1_RPE_PR/cs_mods/mod_RPE.mat\")\n",
    "mod = load_matlab_model('C:/Users/prins/git/Human-GEM/model/Human-GEM.mat') # load HUMAN1 model (this takes really long!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "847fa973-8f40-4e20-a949-167263242fb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No defined compartments in model model. Compartments will be deduced heuristically using regular expressions.\n",
      "Using regular expression found the following compartments:c, e, g, i, l, m, n, r, x\n"
     ]
    }
   ],
   "source": [
    "from cobra.io import load_json_model, save_json_model, load_matlab_model, save_matlab_model, read_sbml_model, write_sbml_model\n",
    "mod_sbml = read_sbml_model('C:/Users/prins/git/Human-GEM/model/Human-GEM.xml') \n",
    "mod_recon3d = load_matlab_model('C:/Users/prins/git/COBRA_project_SP/data/Recon3D_model/Recon3DModel_301.mat') # load Recon3D model (this takes really long!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe564dfe-3fde-46d7-ab09-7aabc3b3dd93",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "l1 = list(set([rxns.subsystem for rxns in mod_recon3d.reactions]) - set([rxns.subsystem for rxns in mod_sbml.reactions]))\n",
    "l2 = list(set([rxns.subsystem for rxns in mod_sbml.reactions]) - set([rxns.subsystem for rxns in mod_recon3d.reactions]))\n",
    "l3 = list(set([rxns.subsystem for rxns in mod_sbml.reactions]).intersection([rxns.subsystem for rxns in mod_recon3d.reactions]))\n",
    "pd.DataFrame(l3).to_clipboard(excel=True, sep=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1041577-cb82-417f-a015-908784fa6463",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAR07108</td>\n",
       "      <td></td>\n",
       "      <td>MAM01374e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAR07110</td>\n",
       "      <td></td>\n",
       "      <td>MAM02556e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAR07112</td>\n",
       "      <td></td>\n",
       "      <td>MAM01296e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAR07114</td>\n",
       "      <td></td>\n",
       "      <td>MAM03044e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAR07116</td>\n",
       "      <td></td>\n",
       "      <td>MAM01403e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660</th>\n",
       "      <td>MAR10088</td>\n",
       "      <td></td>\n",
       "      <td>MAM10031e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661</th>\n",
       "      <td>MAR10089</td>\n",
       "      <td></td>\n",
       "      <td>MAM10032e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1662</th>\n",
       "      <td>MAR10090</td>\n",
       "      <td></td>\n",
       "      <td>MAM10033e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1663</th>\n",
       "      <td>MAR10091</td>\n",
       "      <td></td>\n",
       "      <td>MAM10034e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>MAR10132</td>\n",
       "      <td></td>\n",
       "      <td>MAM10041e &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1665 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0 1                2   3\n",
       "0     MAR07108     MAM01374e <=>   []\n",
       "1     MAR07110     MAM02556e <=>   []\n",
       "2     MAR07112     MAM01296e <=>   []\n",
       "3     MAR07114     MAM03044e <=>   []\n",
       "4     MAR07116     MAM01403e <=>   []\n",
       "...        ... ..             ...  ..\n",
       "1660  MAR10088     MAM10031e <=>   []\n",
       "1661  MAR10089     MAM10032e <=>   []\n",
       "1662  MAR10090     MAM10033e <=>   []\n",
       "1663  MAR10091     MAM10034e <=>   []\n",
       "1664  MAR10132     MAM10041e <=>   []\n",
       "\n",
       "[1665 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_sbml.reactions \\\n",
    " if len(rxns.products) == 0]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16f1ecf7-6483-4b31-87d7-bbed9d797065",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tauro-omega-muricholic acid'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_sbml.metabolites.get_by_any(['MAM10031e'])[0].name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e19f0250-eb22-4609-812a-81450e7397c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAR07108</td>\n",
       "      <td></td>\n",
       "      <td>MAM01374e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627', 'vmhreaction': 'EX_C075...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MAR07110</td>\n",
       "      <td></td>\n",
       "      <td>MAM02556e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627', 'vmhreaction': 'EX_npth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MAR07112</td>\n",
       "      <td></td>\n",
       "      <td>MAM01296e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627', 'vmhreaction': 'EX_afla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAR07114</td>\n",
       "      <td></td>\n",
       "      <td>MAM03044e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627', 'vmhreaction': 'EX_M030...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MAR07116</td>\n",
       "      <td></td>\n",
       "      <td>MAM01403e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627', 'vmhreaction': 'EX_M014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660</th>\n",
       "      <td>MAR10088</td>\n",
       "      <td></td>\n",
       "      <td>MAM10031e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661</th>\n",
       "      <td>MAR10089</td>\n",
       "      <td></td>\n",
       "      <td>MAM10032e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1662</th>\n",
       "      <td>MAR10090</td>\n",
       "      <td></td>\n",
       "      <td>MAM10033e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1663</th>\n",
       "      <td>MAR10091</td>\n",
       "      <td></td>\n",
       "      <td>MAM10034e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>MAR10132</td>\n",
       "      <td></td>\n",
       "      <td>MAM10041e &lt;=&gt;</td>\n",
       "      <td>{'sbo': 'SBO:0000627'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1665 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             0 1                2  \\\n",
       "0     MAR07108     MAM01374e <=>    \n",
       "1     MAR07110     MAM02556e <=>    \n",
       "2     MAR07112     MAM01296e <=>    \n",
       "3     MAR07114     MAM03044e <=>    \n",
       "4     MAR07116     MAM01403e <=>    \n",
       "...        ... ..             ...   \n",
       "1660  MAR10088     MAM10031e <=>    \n",
       "1661  MAR10089     MAM10032e <=>    \n",
       "1662  MAR10090     MAM10033e <=>    \n",
       "1663  MAR10091     MAM10034e <=>    \n",
       "1664  MAR10132     MAM10041e <=>    \n",
       "\n",
       "                                                      3  \n",
       "0     {'sbo': 'SBO:0000627', 'vmhreaction': 'EX_C075...  \n",
       "1     {'sbo': 'SBO:0000627', 'vmhreaction': 'EX_npth...  \n",
       "2     {'sbo': 'SBO:0000627', 'vmhreaction': 'EX_afla...  \n",
       "3     {'sbo': 'SBO:0000627', 'vmhreaction': 'EX_M030...  \n",
       "4     {'sbo': 'SBO:0000627', 'vmhreaction': 'EX_M014...  \n",
       "...                                                 ...  \n",
       "1660                             {'sbo': 'SBO:0000627'}  \n",
       "1661                             {'sbo': 'SBO:0000627'}  \n",
       "1662                             {'sbo': 'SBO:0000627'}  \n",
       "1663                             {'sbo': 'SBO:0000627'}  \n",
       "1664                             {'sbo': 'SBO:0000627'}  \n",
       "\n",
       "[1665 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l=[[rxns.id, rxns.name, rxns.reaction, rxns.annotation]  for rxns in mod_sbml.exchanges]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1ff7bd9a-d8fa-491f-bbd2-1b8e9465512f",
   "metadata": {},
   "outputs": [],
   "source": [
    "l_r3d_trxns = [rxns.name for rxns in mod_recon3d.reactions if rxns.subsystem == 'Transport, extracellular']\n",
    "l_r3d_trxnsid = [rxns.id for rxns in mod_recon3d.reactions if rxns.subsystem == 'Transport, extracellular']\n",
    "pd.DataFrame(l_r3d_trxns, l_r3d_trxnsid).to_clipboard(excel=True, sep=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c13ad1dd-4ac7-4c4a-9436-9691888e2be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "l_r3d_trxns = [[rxns.id, rxns.name, rxns.reaction, rxns.reactants, rxns.products, rxns.compartments, rxns.reversibility] for rxns in mod_recon3d.reactions if rxns.subsystem == 'Transport, extracellular']\n",
    "pd.DataFrame(l_r3d_trxns)\n",
    "pd.DataFrame(l_r3d_trxns).to_clipboard(excel=True, sep=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2390e5db-2b93-439c-a8b5-486b7b3d7a36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Metabolite h2o[m] at 0x1c186b6ddb0>,\n",
       " <Metabolite 1a2425thvitd2[m] at 0x1c186b6dea0>,\n",
       " <Metabolite nadp[m] at 0x1c186b6ded0>]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.reactions[13].products"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "30fb2c10-1e3e-401e-a287-ff5cb2e588f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.reactions.get_by_id('CGLYt3_2_').reversibility == False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "60bafaf8-287f-4ed2-98ef-f0dbb3554efd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['google']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mystring = [\"reddit\", \"google\"]\n",
    "mylist = [\"a\", \"b\", \"c\", \"d\"]\n",
    "[s for s in mystring if not any(x in s for x in mylist)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "441e3d5f-edb0-4dd6-b651-14201f50763e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Transport, extracellular'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.reactions.get_by_id('r0295').subsystem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5b01792c-766f-4ec6-9020-7e7f3b21bb96",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1MNCAMti</td>\n",
       "      <td>N1-Methylnicotinamide Transport</td>\n",
       "      <td>1mncam[c] + atp[c] + h2o[c] --&gt; 1mncam[e] + ad...</td>\n",
       "      <td>[h[c], 1mncam[e], adp[c], pi[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24_25DHVITD2t</td>\n",
       "      <td>24, 25-Dihydroxyvitamin D2 Transport from Cyto...</td>\n",
       "      <td>2425dhvitd2[c] --&gt; 2425dhvitd2[e]</td>\n",
       "      <td>[2425dhvitd2[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24_25DHVITD3t</td>\n",
       "      <td>24, 25-Dihydroxyvitamin D3 Transport from Cyto...</td>\n",
       "      <td>2425dhvitd3[c] --&gt; 2425dhvitd3[e]</td>\n",
       "      <td>[2425dhvitd3[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25HVITD2t</td>\n",
       "      <td>25-Hydroxyvitamin D2 Transport from Cytoplasm</td>\n",
       "      <td>25hvitd2[c] --&gt; 25hvitd2[e]</td>\n",
       "      <td>[25hvitd2[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25HVITD3t</td>\n",
       "      <td>25-Hydroxyvitamin D3 Transport from Cytoplasm</td>\n",
       "      <td>25hvitd3[c] --&gt; 25hvitd3[e]</td>\n",
       "      <td>[25hvitd3[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1359</th>\n",
       "      <td>TUDCA3Sabc</td>\n",
       "      <td>Tauroursodeoxycholic acid 3-sulfate ABC bile a...</td>\n",
       "      <td>atp[c] + h2o[c] + tudca3s[c] --&gt; adp[c] + h[c]...</td>\n",
       "      <td>[h[c], adp[c], pi[c], tudca3s[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360</th>\n",
       "      <td>UCHOLabc</td>\n",
       "      <td>Ursocholic acid ABC bile acid transporter</td>\n",
       "      <td>atp[c] + h2o[c] + uchol[c] --&gt; adp[c] + h[c] +...</td>\n",
       "      <td>[h[c], adp[c], pi[c], uchol[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1361</th>\n",
       "      <td>UCHOLt2</td>\n",
       "      <td>Ursocholic acid transport via sodium cotransport</td>\n",
       "      <td>2.0 na1[e] + uchol[e] --&gt; 2.0 na1[c] + uchol[c]</td>\n",
       "      <td>[na1[c], uchol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362</th>\n",
       "      <td>UDCA3Sabc</td>\n",
       "      <td>Ursodeoxycholic acid 3-sulfate ABC bile acid t...</td>\n",
       "      <td>atp[c] + h2o[c] + udca3s[c] --&gt; adp[c] + h[c] ...</td>\n",
       "      <td>[h[c], adp[c], pi[c], udca3s[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1363</th>\n",
       "      <td>UDCHOLt2</td>\n",
       "      <td>Ursodeoxycholic acid transport via sodium cotr...</td>\n",
       "      <td>HC02194[e] + 2.0 na1[e] --&gt; HC02194[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], HC02194[c]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1364 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                                                  1  \\\n",
       "0          1MNCAMti                    N1-Methylnicotinamide Transport   \n",
       "1     24_25DHVITD2t  24, 25-Dihydroxyvitamin D2 Transport from Cyto...   \n",
       "2     24_25DHVITD3t  24, 25-Dihydroxyvitamin D3 Transport from Cyto...   \n",
       "3         25HVITD2t      25-Hydroxyvitamin D2 Transport from Cytoplasm   \n",
       "4         25HVITD3t      25-Hydroxyvitamin D3 Transport from Cytoplasm   \n",
       "...             ...                                                ...   \n",
       "1359     TUDCA3Sabc  Tauroursodeoxycholic acid 3-sulfate ABC bile a...   \n",
       "1360       UCHOLabc          Ursocholic acid ABC bile acid transporter   \n",
       "1361        UCHOLt2   Ursocholic acid transport via sodium cotransport   \n",
       "1362      UDCA3Sabc  Ursodeoxycholic acid 3-sulfate ABC bile acid t...   \n",
       "1363       UDCHOLt2  Ursodeoxycholic acid transport via sodium cotr...   \n",
       "\n",
       "                                                      2  \\\n",
       "0     1mncam[c] + atp[c] + h2o[c] --> 1mncam[e] + ad...   \n",
       "1                     2425dhvitd2[c] --> 2425dhvitd2[e]   \n",
       "2                     2425dhvitd3[c] --> 2425dhvitd3[e]   \n",
       "3                           25hvitd2[c] --> 25hvitd2[e]   \n",
       "4                           25hvitd3[c] --> 25hvitd3[e]   \n",
       "...                                                 ...   \n",
       "1359  atp[c] + h2o[c] + tudca3s[c] --> adp[c] + h[c]...   \n",
       "1360  atp[c] + h2o[c] + uchol[c] --> adp[c] + h[c] +...   \n",
       "1361    2.0 na1[e] + uchol[e] --> 2.0 na1[c] + uchol[c]   \n",
       "1362  atp[c] + h2o[c] + udca3s[c] --> adp[c] + h[c] ...   \n",
       "1363  HC02194[e] + 2.0 na1[e] --> HC02194[c] + 2.0 n...   \n",
       "\n",
       "                                      3  \n",
       "0      [h[c], 1mncam[e], adp[c], pi[c]]  \n",
       "1                      [2425dhvitd2[e]]  \n",
       "2                      [2425dhvitd3[e]]  \n",
       "3                         [25hvitd2[e]]  \n",
       "4                         [25hvitd3[e]]  \n",
       "...                                 ...  \n",
       "1359  [h[c], adp[c], pi[c], tudca3s[e]]  \n",
       "1360    [h[c], adp[c], pi[c], uchol[e]]  \n",
       "1361                 [na1[c], uchol[c]]  \n",
       "1362   [h[c], adp[c], pi[c], udca3s[e]]  \n",
       "1363               [na1[c], HC02194[c]]  \n",
       "\n",
       "[1364 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.compartments\n",
    "comp = [\"[e]\", \"[c]\"]\n",
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if any(x in str(rxns.products) for x in comp)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b272c809-f980-458a-82d0-2e5c08c4492b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTMLTe</td>\n",
       "      <td>G-Glutamyltransferase, Extracellular</td>\n",
       "      <td>ala_L[c] + gthrd[c] --&gt; cgly[e] + gluala[e]</td>\n",
       "      <td>Glutathione metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[ala_L[c], gthrd[c]]</td>\n",
       "      <td>[cgly[e], gluala[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HAS1</td>\n",
       "      <td>Hyaluronan Synthase</td>\n",
       "      <td>uacgam[c] + udpglcur[c] --&gt; 2.0 h[c] + ha_pre1...</td>\n",
       "      <td>Hyaluronan metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[uacgam[c], udpglcur[c]]</td>\n",
       "      <td>[h[c], udp[c], ha_pre1[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HAS2</td>\n",
       "      <td>Hyaluronan Synthase</td>\n",
       "      <td>ha_pre1[e] + uacgam[c] + udpglcur[c] --&gt; 2.0 h...</td>\n",
       "      <td>Hyaluronan metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[ha_pre1[e], uacgam[c], udpglcur[c]]</td>\n",
       "      <td>[h[c], udp[c], ha[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RETNGLCt2</td>\n",
       "      <td>Retinoyl Glucuronide Efflux (13-Cis)</td>\n",
       "      <td>13_cis_retnglc[c] --&gt; 13_cis_retnglc[e]</td>\n",
       "      <td>Transport, endoplasmic reticular</td>\n",
       "      <td>False</td>\n",
       "      <td>[13_cis_retnglc[c]]</td>\n",
       "      <td>[13_cis_retnglc[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLDt</td>\n",
       "      <td>L-Sulfolactate Transport (Cytosol to Extracell...</td>\n",
       "      <td>sl_L[c] --&gt; sl_L[e]</td>\n",
       "      <td>Methionine and cysteine metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[sl_L[c]]</td>\n",
       "      <td>[sl_L[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>SULPACMPtev</td>\n",
       "      <td>efflux of acetaminophen-sulphate into hepatic ...</td>\n",
       "      <td>atp[c] + h2o[c] + sulpacmp[c] --&gt; adp[c] + h[c...</td>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[h2o[c], atp[c], sulpacmp[c]]</td>\n",
       "      <td>[h[c], adp[c], pi[c], sulpacmp[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>TACRtu</td>\n",
       "      <td>efflux of tacrolimus into intestinal lumen by ...</td>\n",
       "      <td>atp[c] + h2o[c] + tacr[c] --&gt; adp[c] + h[c] + ...</td>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[h2o[c], atp[c], tacr[c]]</td>\n",
       "      <td>[h[c], adp[c], pi[c], tacr[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>TRIPVSteb</td>\n",
       "      <td>efflux of triol metabolite of pravastatin into...</td>\n",
       "      <td>atp[c] + h2o[c] + tripvs[c] --&gt; adp[c] + h[c] ...</td>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[h2o[c], atp[c], tripvs[c]]</td>\n",
       "      <td>[h[c], adp[c], pi[c], tripvs[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>TSACGLUCtev</td>\n",
       "      <td>efflux of thiomethyl-sulphoxide-acetaminophen-...</td>\n",
       "      <td>atp[c] + h2o[c] + tsacmgluc[c] --&gt; adp[c] + h[...</td>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[h2o[c], atp[c], tsacmgluc[c]]</td>\n",
       "      <td>[h[c], adp[c], pi[c], tsacmgluc[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>TSACMSULtev</td>\n",
       "      <td>efflux of thiomethyl-sulphoxide-acetaminophen-...</td>\n",
       "      <td>atp[c] + h2o[c] + tsacmsul[c] --&gt; adp[c] + h[c...</td>\n",
       "      <td>Drug metabolism</td>\n",
       "      <td>False</td>\n",
       "      <td>[h2o[c], atp[c], tsacmsul[c]]</td>\n",
       "      <td>[h[c], adp[c], pi[c], tsacmsul[e]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>145 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0                                                  1  \\\n",
       "0         GTMLTe               G-Glutamyltransferase, Extracellular   \n",
       "1           HAS1                                Hyaluronan Synthase   \n",
       "2           HAS2                                Hyaluronan Synthase   \n",
       "3      RETNGLCt2               Retinoyl Glucuronide Efflux (13-Cis)   \n",
       "4           SLDt  L-Sulfolactate Transport (Cytosol to Extracell...   \n",
       "..           ...                                                ...   \n",
       "140  SULPACMPtev  efflux of acetaminophen-sulphate into hepatic ...   \n",
       "141       TACRtu  efflux of tacrolimus into intestinal lumen by ...   \n",
       "142    TRIPVSteb  efflux of triol metabolite of pravastatin into...   \n",
       "143  TSACGLUCtev  efflux of thiomethyl-sulphoxide-acetaminophen-...   \n",
       "144  TSACMSULtev  efflux of thiomethyl-sulphoxide-acetaminophen-...   \n",
       "\n",
       "                                                     2  \\\n",
       "0          ala_L[c] + gthrd[c] --> cgly[e] + gluala[e]   \n",
       "1    uacgam[c] + udpglcur[c] --> 2.0 h[c] + ha_pre1...   \n",
       "2    ha_pre1[e] + uacgam[c] + udpglcur[c] --> 2.0 h...   \n",
       "3              13_cis_retnglc[c] --> 13_cis_retnglc[e]   \n",
       "4                                  sl_L[c] --> sl_L[e]   \n",
       "..                                                 ...   \n",
       "140  atp[c] + h2o[c] + sulpacmp[c] --> adp[c] + h[c...   \n",
       "141  atp[c] + h2o[c] + tacr[c] --> adp[c] + h[c] + ...   \n",
       "142  atp[c] + h2o[c] + tripvs[c] --> adp[c] + h[c] ...   \n",
       "143  atp[c] + h2o[c] + tsacmgluc[c] --> adp[c] + h[...   \n",
       "144  atp[c] + h2o[c] + tsacmsul[c] --> adp[c] + h[c...   \n",
       "\n",
       "                                      3      4  \\\n",
       "0                Glutathione metabolism  False   \n",
       "1                 Hyaluronan metabolism  False   \n",
       "2                 Hyaluronan metabolism  False   \n",
       "3      Transport, endoplasmic reticular  False   \n",
       "4    Methionine and cysteine metabolism  False   \n",
       "..                                  ...    ...   \n",
       "140                     Drug metabolism  False   \n",
       "141                     Drug metabolism  False   \n",
       "142                     Drug metabolism  False   \n",
       "143                     Drug metabolism  False   \n",
       "144                     Drug metabolism  False   \n",
       "\n",
       "                                        5                                    6  \n",
       "0                    [ala_L[c], gthrd[c]]                 [cgly[e], gluala[e]]  \n",
       "1                [uacgam[c], udpglcur[c]]           [h[c], udp[c], ha_pre1[e]]  \n",
       "2    [ha_pre1[e], uacgam[c], udpglcur[c]]                [h[c], udp[c], ha[e]]  \n",
       "3                     [13_cis_retnglc[c]]                  [13_cis_retnglc[e]]  \n",
       "4                               [sl_L[c]]                            [sl_L[e]]  \n",
       "..                                    ...                                  ...  \n",
       "140         [h2o[c], atp[c], sulpacmp[c]]   [h[c], adp[c], pi[c], sulpacmp[e]]  \n",
       "141             [h2o[c], atp[c], tacr[c]]       [h[c], adp[c], pi[c], tacr[e]]  \n",
       "142           [h2o[c], atp[c], tripvs[c]]     [h[c], adp[c], pi[c], tripvs[e]]  \n",
       "143        [h2o[c], atp[c], tsacmgluc[c]]  [h[c], adp[c], pi[c], tsacmgluc[e]]  \n",
       "144         [h2o[c], atp[c], tsacmsul[c]]   [h[c], adp[c], pi[c], tsacmsul[e]]  \n",
       "\n",
       "[145 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.compartments\n",
    "comp = [\"[c]\", \"[l]\", \"[m]\", \"[r]\", \"[x]\", \"[n]\", \"[g]\",\"[i]\"]\n",
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.subsystem, rxns.reversibility, rxns.reactants, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if any(x in str(rxns.reactants) for x in comp)\\\n",
    " if '[e]' in str(rxns.products)\\\n",
    " if not rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "048dad0b-83e9-4548-be20-87df7b291590",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1MNCAMti</td>\n",
       "      <td>N1-Methylnicotinamide Transport</td>\n",
       "      <td>1mncam[c] + atp[c] + h2o[c] --&gt; 1mncam[e] + ad...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>24_25DHVITD2t</td>\n",
       "      <td>24, 25-Dihydroxyvitamin D2 Transport from Cyto...</td>\n",
       "      <td>2425dhvitd2[c] --&gt; 2425dhvitd2[e]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>24_25DHVITD3t</td>\n",
       "      <td>24, 25-Dihydroxyvitamin D3 Transport from Cyto...</td>\n",
       "      <td>2425dhvitd3[c] --&gt; 2425dhvitd3[e]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25HVITD2t</td>\n",
       "      <td>25-Hydroxyvitamin D2 Transport from Cytoplasm</td>\n",
       "      <td>25hvitd2[c] --&gt; 25hvitd2[e]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>25HVITD3t</td>\n",
       "      <td>25-Hydroxyvitamin D3 Transport from Cytoplasm</td>\n",
       "      <td>25hvitd3[c] --&gt; 25hvitd3[e]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1373</th>\n",
       "      <td>TUDCA3Sabc</td>\n",
       "      <td>Tauroursodeoxycholic acid 3-sulfate ABC bile a...</td>\n",
       "      <td>atp[c] + h2o[c] + tudca3s[c] --&gt; adp[c] + h[c]...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1374</th>\n",
       "      <td>UCHOLabc</td>\n",
       "      <td>Ursocholic acid ABC bile acid transporter</td>\n",
       "      <td>atp[c] + h2o[c] + uchol[c] --&gt; adp[c] + h[c] +...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1375</th>\n",
       "      <td>UCHOLt2</td>\n",
       "      <td>Ursocholic acid transport via sodium cotransport</td>\n",
       "      <td>2.0 na1[e] + uchol[e] --&gt; 2.0 na1[c] + uchol[c]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1376</th>\n",
       "      <td>UDCA3Sabc</td>\n",
       "      <td>Ursodeoxycholic acid 3-sulfate ABC bile acid t...</td>\n",
       "      <td>atp[c] + h2o[c] + udca3s[c] --&gt; adp[c] + h[c] ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1377</th>\n",
       "      <td>UDCHOLt2</td>\n",
       "      <td>Ursodeoxycholic acid transport via sodium cotr...</td>\n",
       "      <td>HC02194[e] + 2.0 na1[e] --&gt; HC02194[c] + 2.0 n...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1378 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  0                                                  1  \\\n",
       "0          1MNCAMti                    N1-Methylnicotinamide Transport   \n",
       "1     24_25DHVITD2t  24, 25-Dihydroxyvitamin D2 Transport from Cyto...   \n",
       "2     24_25DHVITD3t  24, 25-Dihydroxyvitamin D3 Transport from Cyto...   \n",
       "3         25HVITD2t      25-Hydroxyvitamin D2 Transport from Cytoplasm   \n",
       "4         25HVITD3t      25-Hydroxyvitamin D3 Transport from Cytoplasm   \n",
       "...             ...                                                ...   \n",
       "1373     TUDCA3Sabc  Tauroursodeoxycholic acid 3-sulfate ABC bile a...   \n",
       "1374       UCHOLabc          Ursocholic acid ABC bile acid transporter   \n",
       "1375        UCHOLt2   Ursocholic acid transport via sodium cotransport   \n",
       "1376      UDCA3Sabc  Ursodeoxycholic acid 3-sulfate ABC bile acid t...   \n",
       "1377       UDCHOLt2  Ursodeoxycholic acid transport via sodium cotr...   \n",
       "\n",
       "                                                      2  \n",
       "0     1mncam[c] + atp[c] + h2o[c] --> 1mncam[e] + ad...  \n",
       "1                     2425dhvitd2[c] --> 2425dhvitd2[e]  \n",
       "2                     2425dhvitd3[c] --> 2425dhvitd3[e]  \n",
       "3                           25hvitd2[c] --> 25hvitd2[e]  \n",
       "4                           25hvitd3[c] --> 25hvitd3[e]  \n",
       "...                                                 ...  \n",
       "1373  atp[c] + h2o[c] + tudca3s[c] --> adp[c] + h[c]...  \n",
       "1374  atp[c] + h2o[c] + uchol[c] --> adp[c] + h[c] +...  \n",
       "1375    2.0 na1[e] + uchol[e] --> 2.0 na1[c] + uchol[c]  \n",
       "1376  atp[c] + h2o[c] + udca3s[c] --> adp[c] + h[c] ...  \n",
       "1377  HC02194[e] + 2.0 na1[e] --> HC02194[c] + 2.0 n...  \n",
       "\n",
       "[1378 rows x 3 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction] for rxns in mod_recon3d.reactions \\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0f96bc0a-923f-4a6f-a580-118a4126d6b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GLCMter</td>\n",
       "      <td>Glucose Transport via Membrane Vesicle</td>\n",
       "      <td>glc_D[r] --&gt; glc_D[e]</td>\n",
       "      <td>[glc_D[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>r0813</td>\n",
       "      <td>ATP-Binding Cassette (ABC) Tcdb:3.A.1.208.9</td>\n",
       "      <td>atp[r] + bildglcur[r] + h2o[r] --&gt; adp[r] + bi...</td>\n",
       "      <td>[adp[r], bildglcur[e], h[r], pi[r]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>r0924</td>\n",
       "      <td>ATP-Binding Cassette (ABC) Tcdb:3.A.1.204.5</td>\n",
       "      <td>atp[r] + chsterol[r] + h2o[r] --&gt; adp[r] + chs...</td>\n",
       "      <td>[adp[r], h[r], pi[r], chsterol[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>r1017</td>\n",
       "      <td>Vesicular Transport</td>\n",
       "      <td>atp[x] + h2o[x] + tchola[x] --&gt; adp[x] + h[x] ...</td>\n",
       "      <td>[h[x], adp[x], pi[x], tchola[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>r1029</td>\n",
       "      <td>Vesicular Transport</td>\n",
       "      <td>atp[x] + h2o[x] + tdchola[x] --&gt; adp[x] + h[x]...</td>\n",
       "      <td>[h[x], adp[x], pi[x], tdchola[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>LEUKTRB4WOHtr</td>\n",
       "      <td>Transport of W-Hydroxyl Leukotriene B4, Intrac...</td>\n",
       "      <td>atp[r] + h2o[r] + leuktrB4woh[r] --&gt; adp[r] + ...</td>\n",
       "      <td>[adp[r], h[r], pi[r], leuktrB4woh[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HMR_1897</td>\n",
       "      <td>HMR_1897</td>\n",
       "      <td>atp[r] + bilglcur[r] + h2o[r] --&gt; adp[r] + bil...</td>\n",
       "      <td>[adp[r], bilglcur[e], h[r], pi[r]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0                                                  1  \\\n",
       "0        GLCMter             Glucose Transport via Membrane Vesicle   \n",
       "1          r0813        ATP-Binding Cassette (ABC) Tcdb:3.A.1.208.9   \n",
       "2          r0924        ATP-Binding Cassette (ABC) Tcdb:3.A.1.204.5   \n",
       "3          r1017                                Vesicular Transport   \n",
       "4          r1029                                Vesicular Transport   \n",
       "5  LEUKTRB4WOHtr  Transport of W-Hydroxyl Leukotriene B4, Intrac...   \n",
       "6       HMR_1897                                           HMR_1897   \n",
       "\n",
       "                                                   2  \\\n",
       "0                              glc_D[r] --> glc_D[e]   \n",
       "1  atp[r] + bildglcur[r] + h2o[r] --> adp[r] + bi...   \n",
       "2  atp[r] + chsterol[r] + h2o[r] --> adp[r] + chs...   \n",
       "3  atp[x] + h2o[x] + tchola[x] --> adp[x] + h[x] ...   \n",
       "4  atp[x] + h2o[x] + tdchola[x] --> adp[x] + h[x]...   \n",
       "5  atp[r] + h2o[r] + leuktrB4woh[r] --> adp[r] + ...   \n",
       "6  atp[r] + bilglcur[r] + h2o[r] --> adp[r] + bil...   \n",
       "\n",
       "                                       3  \n",
       "0                             [glc_D[e]]  \n",
       "1    [adp[r], bildglcur[e], h[r], pi[r]]  \n",
       "2     [adp[r], h[r], pi[r], chsterol[e]]  \n",
       "3       [h[x], adp[x], pi[x], tchola[e]]  \n",
       "4      [h[x], adp[x], pi[x], tdchola[e]]  \n",
       "5  [adp[r], h[r], pi[r], leuktrB4woh[c]]  \n",
       "6     [adp[r], bilglcur[e], h[r], pi[r]]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.compartments\n",
    "comp = [\"[r]\", \"[x]\", \"[n]\", \"[i]\"]\n",
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if any(x in str(rxns.reaction) for x in comp)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0a6f5e94-8c77-40c3-8e4f-ca62dfa6f1ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHOLESTle</td>\n",
       "      <td>Hydrolysis of Cholesterol Ester by Cholesterol...</td>\n",
       "      <td>h2o[e] + xolest2_hs[e] --&gt; Rtotal[e] + chstero...</td>\n",
       "      <td>[h[e], chsterol[e], Rtotal[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYSAMOe</td>\n",
       "      <td>Cysteamine:Oxygen Oxidoreductase, Extracellular</td>\n",
       "      <td>cysam[e] + o2[e] --&gt; h[e] + hyptaur[e]</td>\n",
       "      <td>[h[e], hyptaur[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FMNALKPle</td>\n",
       "      <td>Acid Phosphatase (FMN), Extracellular</td>\n",
       "      <td>fmn[e] + h2o[e] --&gt; pi[e] + ribflv[e]</td>\n",
       "      <td>[pi[e], ribflv[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PNTEHe</td>\n",
       "      <td>(R)-Pantetheine Amidohydrolase</td>\n",
       "      <td>h2o[e] + ptth[e] --&gt; cysam[e] + pnto_R[e]</td>\n",
       "      <td>[pnto_R[e], cysam[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PTPATe</td>\n",
       "      <td>Phosphopantetheine Adenylyltransferase, Extrac...</td>\n",
       "      <td>atp[e] + h[e] + pan4p[e] --&gt; dpcoa[e] + ppi[e]</td>\n",
       "      <td>[ppi[e], dpcoa[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMHISTAe</td>\n",
       "      <td>S-Adenosyl-L-Methionine:Histamine N-Tele-Methy...</td>\n",
       "      <td>amet[e] + hista[e] --&gt; ahcys[e] + h[e] + mhist...</td>\n",
       "      <td>[h[e], ahcys[e], mhista[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HMR_0008</td>\n",
       "      <td>Lipoprotein Lipase</td>\n",
       "      <td>M00241[e] + h2o[e] --&gt; M00510[e] + M01820[e]</td>\n",
       "      <td>[M01820[e], M00510[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HMR_0015</td>\n",
       "      <td>HMR_0015</td>\n",
       "      <td>M02560[e] --&gt; 0.0001 CE2510[e] + 0.0048 M00003...</td>\n",
       "      <td>[adrn[e], arachd[e], arach[e], clpnd[e], crvnc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HMR_0016</td>\n",
       "      <td>HMR_0016</td>\n",
       "      <td>0.0046 CE2510[e] + 0.0001 M00003[e] + 0.0023 M...</td>\n",
       "      <td>[M02561[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HMR_0017</td>\n",
       "      <td>HMR_0017</td>\n",
       "      <td>M02909[e] --&gt; 0.125 C01601[e] + 0.125 M02108[e...</td>\n",
       "      <td>[octa[e], dca[e], but[e], C01601[e], M02108[e]...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                                                  1  \\\n",
       "0  CHOLESTle  Hydrolysis of Cholesterol Ester by Cholesterol...   \n",
       "1    CYSAMOe    Cysteamine:Oxygen Oxidoreductase, Extracellular   \n",
       "2  FMNALKPle              Acid Phosphatase (FMN), Extracellular   \n",
       "3     PNTEHe                     (R)-Pantetheine Amidohydrolase   \n",
       "4     PTPATe  Phosphopantetheine Adenylyltransferase, Extrac...   \n",
       "5  SAMHISTAe  S-Adenosyl-L-Methionine:Histamine N-Tele-Methy...   \n",
       "6   HMR_0008                                 Lipoprotein Lipase   \n",
       "7   HMR_0015                                           HMR_0015   \n",
       "8   HMR_0016                                           HMR_0016   \n",
       "9   HMR_0017                                           HMR_0017   \n",
       "\n",
       "                                                   2  \\\n",
       "0  h2o[e] + xolest2_hs[e] --> Rtotal[e] + chstero...   \n",
       "1             cysam[e] + o2[e] --> h[e] + hyptaur[e]   \n",
       "2              fmn[e] + h2o[e] --> pi[e] + ribflv[e]   \n",
       "3          h2o[e] + ptth[e] --> cysam[e] + pnto_R[e]   \n",
       "4     atp[e] + h[e] + pan4p[e] --> dpcoa[e] + ppi[e]   \n",
       "5  amet[e] + hista[e] --> ahcys[e] + h[e] + mhist...   \n",
       "6       M00241[e] + h2o[e] --> M00510[e] + M01820[e]   \n",
       "7  M02560[e] --> 0.0001 CE2510[e] + 0.0048 M00003...   \n",
       "8  0.0046 CE2510[e] + 0.0001 M00003[e] + 0.0023 M...   \n",
       "9  M02909[e] --> 0.125 C01601[e] + 0.125 M02108[e...   \n",
       "\n",
       "                                                   3  \n",
       "0                     [h[e], chsterol[e], Rtotal[e]]  \n",
       "1                                 [h[e], hyptaur[e]]  \n",
       "2                                 [pi[e], ribflv[e]]  \n",
       "3                              [pnto_R[e], cysam[e]]  \n",
       "4                                 [ppi[e], dpcoa[e]]  \n",
       "5                        [h[e], ahcys[e], mhista[e]]  \n",
       "6                             [M01820[e], M00510[e]]  \n",
       "7  [adrn[e], arachd[e], arach[e], clpnd[e], crvnc...  \n",
       "8                                        [M02561[e]]  \n",
       "9  [octa[e], dca[e], but[e], C01601[e], M02108[e]...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.compartments\n",
    "comp = [\"[c]\", \"[l]\", \"[m]\", \"[r]\", \"[x]\", \"[n]\", \"[g]\",\"[i]\"]\n",
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if not any(x in str(rxns.reaction) for x in comp)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "996379f8-5cdd-4d7f-8432-fa611d7aba3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHOLESTle</td>\n",
       "      <td>Hydrolysis of Cholesterol Ester by Cholesterol...</td>\n",
       "      <td>h2o[e] + xolest2_hs[e] --&gt; Rtotal[e] + chstero...</td>\n",
       "      <td>[h[e], chsterol[e], Rtotal[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYSAMOe</td>\n",
       "      <td>Cysteamine:Oxygen Oxidoreductase, Extracellular</td>\n",
       "      <td>cysam[e] + o2[e] --&gt; h[e] + hyptaur[e]</td>\n",
       "      <td>[h[e], hyptaur[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FMNALKPle</td>\n",
       "      <td>Acid Phosphatase (FMN), Extracellular</td>\n",
       "      <td>fmn[e] + h2o[e] --&gt; pi[e] + ribflv[e]</td>\n",
       "      <td>[pi[e], ribflv[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PNTEHe</td>\n",
       "      <td>(R)-Pantetheine Amidohydrolase</td>\n",
       "      <td>h2o[e] + ptth[e] --&gt; cysam[e] + pnto_R[e]</td>\n",
       "      <td>[pnto_R[e], cysam[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PTPATe</td>\n",
       "      <td>Phosphopantetheine Adenylyltransferase, Extrac...</td>\n",
       "      <td>atp[e] + h[e] + pan4p[e] --&gt; dpcoa[e] + ppi[e]</td>\n",
       "      <td>[ppi[e], dpcoa[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMHISTAe</td>\n",
       "      <td>S-Adenosyl-L-Methionine:Histamine N-Tele-Methy...</td>\n",
       "      <td>amet[e] + hista[e] --&gt; ahcys[e] + h[e] + mhist...</td>\n",
       "      <td>[h[e], ahcys[e], mhista[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HMR_0008</td>\n",
       "      <td>Lipoprotein Lipase</td>\n",
       "      <td>M00241[e] + h2o[e] --&gt; M00510[e] + M01820[e]</td>\n",
       "      <td>[M01820[e], M00510[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HMR_0015</td>\n",
       "      <td>HMR_0015</td>\n",
       "      <td>M02560[e] --&gt; 0.0001 CE2510[e] + 0.0048 M00003...</td>\n",
       "      <td>[adrn[e], arachd[e], arach[e], clpnd[e], crvnc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HMR_0016</td>\n",
       "      <td>HMR_0016</td>\n",
       "      <td>0.0046 CE2510[e] + 0.0001 M00003[e] + 0.0023 M...</td>\n",
       "      <td>[M02561[e]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HMR_0017</td>\n",
       "      <td>HMR_0017</td>\n",
       "      <td>M02909[e] --&gt; 0.125 C01601[e] + 0.125 M02108[e...</td>\n",
       "      <td>[octa[e], dca[e], but[e], C01601[e], M02108[e]...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           0                                                  1  \\\n",
       "0  CHOLESTle  Hydrolysis of Cholesterol Ester by Cholesterol...   \n",
       "1    CYSAMOe    Cysteamine:Oxygen Oxidoreductase, Extracellular   \n",
       "2  FMNALKPle              Acid Phosphatase (FMN), Extracellular   \n",
       "3     PNTEHe                     (R)-Pantetheine Amidohydrolase   \n",
       "4     PTPATe  Phosphopantetheine Adenylyltransferase, Extrac...   \n",
       "5  SAMHISTAe  S-Adenosyl-L-Methionine:Histamine N-Tele-Methy...   \n",
       "6   HMR_0008                                 Lipoprotein Lipase   \n",
       "7   HMR_0015                                           HMR_0015   \n",
       "8   HMR_0016                                           HMR_0016   \n",
       "9   HMR_0017                                           HMR_0017   \n",
       "\n",
       "                                                   2  \\\n",
       "0  h2o[e] + xolest2_hs[e] --> Rtotal[e] + chstero...   \n",
       "1             cysam[e] + o2[e] --> h[e] + hyptaur[e]   \n",
       "2              fmn[e] + h2o[e] --> pi[e] + ribflv[e]   \n",
       "3          h2o[e] + ptth[e] --> cysam[e] + pnto_R[e]   \n",
       "4     atp[e] + h[e] + pan4p[e] --> dpcoa[e] + ppi[e]   \n",
       "5  amet[e] + hista[e] --> ahcys[e] + h[e] + mhist...   \n",
       "6       M00241[e] + h2o[e] --> M00510[e] + M01820[e]   \n",
       "7  M02560[e] --> 0.0001 CE2510[e] + 0.0048 M00003...   \n",
       "8  0.0046 CE2510[e] + 0.0001 M00003[e] + 0.0023 M...   \n",
       "9  M02909[e] --> 0.125 C01601[e] + 0.125 M02108[e...   \n",
       "\n",
       "                                                   3  \n",
       "0                     [h[e], chsterol[e], Rtotal[e]]  \n",
       "1                                 [h[e], hyptaur[e]]  \n",
       "2                                 [pi[e], ribflv[e]]  \n",
       "3                              [pnto_R[e], cysam[e]]  \n",
       "4                                 [ppi[e], dpcoa[e]]  \n",
       "5                        [h[e], ahcys[e], mhista[e]]  \n",
       "6                             [M01820[e], M00510[e]]  \n",
       "7  [adrn[e], arachd[e], arach[e], clpnd[e], crvnc...  \n",
       "8                                        [M02561[e]]  \n",
       "9  [octa[e], dca[e], but[e], C01601[e], M02108[e]...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.compartments\n",
    "comp = [\"[c]\", \"[l]\", \"[m]\", \"[r]\", \"[x]\", \"[n]\", \"[g]\",\"[i]\"]\n",
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if not any(x in str(rxns.reaction) for x in comp)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "61a5b7f3-5b62-435d-ad6c-2cb9b1dcd55b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADNt4</td>\n",
       "      <td>Adenosine Transport (Na/Adn Symport)</td>\n",
       "      <td>adn[e] + na1[e] --&gt; adn[c] + na1[c]</td>\n",
       "      <td>[na1[c], adn[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADPRIBt</td>\n",
       "      <td>ADPribose Transport</td>\n",
       "      <td>adprib[e] --&gt; adprib[c]</td>\n",
       "      <td>[adprib[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALAt4</td>\n",
       "      <td>Alanine-Sodium Symporter</td>\n",
       "      <td>ala_L[e] + na1[e] --&gt; ala_L[c] + na1[c]</td>\n",
       "      <td>[na1[c], ala_L[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AQCOBALt</td>\n",
       "      <td>Aquacob-(III)-Alamin Uptake</td>\n",
       "      <td>aqcobal[e] --&gt; aqcobal[c]</td>\n",
       "      <td>[aqcobal[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARGt4</td>\n",
       "      <td>L-Arginine Transport in via Sodium Symport</td>\n",
       "      <td>arg_L[e] + na1[e] --&gt; arg_L[c] + na1[c]</td>\n",
       "      <td>[na1[c], arg_L[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ICDCHOLt2</td>\n",
       "      <td>Isochenodeoxycholic acid transport via sodium ...</td>\n",
       "      <td>icdchol[e] + 2.0 na1[e] --&gt; icdchol[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], icdchol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>ISOCHOLt2</td>\n",
       "      <td>Isocholic acid transport via sodium cotransport</td>\n",
       "      <td>isochol[e] + 2.0 na1[e] --&gt; isochol[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], isochol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>THYOCHOLt2</td>\n",
       "      <td>taurohyocholic acid transport via sodium cotra...</td>\n",
       "      <td>2.0 na1[e] + thyochol[e] --&gt; 2.0 na1[c] + thyo...</td>\n",
       "      <td>[na1[c], thyochol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>UCHOLt2</td>\n",
       "      <td>Ursocholic acid transport via sodium cotransport</td>\n",
       "      <td>2.0 na1[e] + uchol[e] --&gt; 2.0 na1[c] + uchol[c]</td>\n",
       "      <td>[na1[c], uchol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>UDCHOLt2</td>\n",
       "      <td>Ursodeoxycholic acid transport via sodium cotr...</td>\n",
       "      <td>HC02194[e] + 2.0 na1[e] --&gt; HC02194[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], HC02194[c]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>315 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                                                  1  \\\n",
       "0         ADNt4               Adenosine Transport (Na/Adn Symport)   \n",
       "1       ADPRIBt                                ADPribose Transport   \n",
       "2         ALAt4                           Alanine-Sodium Symporter   \n",
       "3      AQCOBALt                        Aquacob-(III)-Alamin Uptake   \n",
       "4         ARGt4         L-Arginine Transport in via Sodium Symport   \n",
       "..          ...                                                ...   \n",
       "310   ICDCHOLt2  Isochenodeoxycholic acid transport via sodium ...   \n",
       "311   ISOCHOLt2    Isocholic acid transport via sodium cotransport   \n",
       "312  THYOCHOLt2  taurohyocholic acid transport via sodium cotra...   \n",
       "313     UCHOLt2   Ursocholic acid transport via sodium cotransport   \n",
       "314    UDCHOLt2  Ursodeoxycholic acid transport via sodium cotr...   \n",
       "\n",
       "                                                     2                      3  \n",
       "0                  adn[e] + na1[e] --> adn[c] + na1[c]       [na1[c], adn[c]]  \n",
       "1                              adprib[e] --> adprib[c]            [adprib[c]]  \n",
       "2              ala_L[e] + na1[e] --> ala_L[c] + na1[c]     [na1[c], ala_L[c]]  \n",
       "3                            aqcobal[e] --> aqcobal[c]           [aqcobal[c]]  \n",
       "4              arg_L[e] + na1[e] --> arg_L[c] + na1[c]     [na1[c], arg_L[c]]  \n",
       "..                                                 ...                    ...  \n",
       "310  icdchol[e] + 2.0 na1[e] --> icdchol[c] + 2.0 n...   [na1[c], icdchol[c]]  \n",
       "311  isochol[e] + 2.0 na1[e] --> isochol[c] + 2.0 n...   [na1[c], isochol[c]]  \n",
       "312  2.0 na1[e] + thyochol[e] --> 2.0 na1[c] + thyo...  [na1[c], thyochol[c]]  \n",
       "313    2.0 na1[e] + uchol[e] --> 2.0 na1[c] + uchol[c]     [na1[c], uchol[c]]  \n",
       "314  HC02194[e] + 2.0 na1[e] --> HC02194[c] + 2.0 n...   [na1[c], HC02194[c]]  \n",
       "\n",
       "[315 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if '[e]' not in str(rxns.products)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a81fcb98-f1bb-45ad-90b2-bb6b27aed608",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADNt4</td>\n",
       "      <td>Adenosine Transport (Na/Adn Symport)</td>\n",
       "      <td>adn[e] + na1[e] --&gt; adn[c] + na1[c]</td>\n",
       "      <td>[na1[c], adn[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADPRIBt</td>\n",
       "      <td>ADPribose Transport</td>\n",
       "      <td>adprib[e] --&gt; adprib[c]</td>\n",
       "      <td>[adprib[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ALAt4</td>\n",
       "      <td>Alanine-Sodium Symporter</td>\n",
       "      <td>ala_L[e] + na1[e] --&gt; ala_L[c] + na1[c]</td>\n",
       "      <td>[na1[c], ala_L[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AQCOBALt</td>\n",
       "      <td>Aquacob-(III)-Alamin Uptake</td>\n",
       "      <td>aqcobal[e] --&gt; aqcobal[c]</td>\n",
       "      <td>[aqcobal[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARGt4</td>\n",
       "      <td>L-Arginine Transport in via Sodium Symport</td>\n",
       "      <td>arg_L[e] + na1[e] --&gt; arg_L[c] + na1[c]</td>\n",
       "      <td>[na1[c], arg_L[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>ICDCHOLt2</td>\n",
       "      <td>Isochenodeoxycholic acid transport via sodium ...</td>\n",
       "      <td>icdchol[e] + 2.0 na1[e] --&gt; icdchol[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], icdchol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>ISOCHOLt2</td>\n",
       "      <td>Isocholic acid transport via sodium cotransport</td>\n",
       "      <td>isochol[e] + 2.0 na1[e] --&gt; isochol[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], isochol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>THYOCHOLt2</td>\n",
       "      <td>taurohyocholic acid transport via sodium cotra...</td>\n",
       "      <td>2.0 na1[e] + thyochol[e] --&gt; 2.0 na1[c] + thyo...</td>\n",
       "      <td>[na1[c], thyochol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>UCHOLt2</td>\n",
       "      <td>Ursocholic acid transport via sodium cotransport</td>\n",
       "      <td>2.0 na1[e] + uchol[e] --&gt; 2.0 na1[c] + uchol[c]</td>\n",
       "      <td>[na1[c], uchol[c]]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>UDCHOLt2</td>\n",
       "      <td>Ursodeoxycholic acid transport via sodium cotr...</td>\n",
       "      <td>HC02194[e] + 2.0 na1[e] --&gt; HC02194[c] + 2.0 n...</td>\n",
       "      <td>[na1[c], HC02194[c]]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>315 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              0                                                  1  \\\n",
       "0         ADNt4               Adenosine Transport (Na/Adn Symport)   \n",
       "1       ADPRIBt                                ADPribose Transport   \n",
       "2         ALAt4                           Alanine-Sodium Symporter   \n",
       "3      AQCOBALt                        Aquacob-(III)-Alamin Uptake   \n",
       "4         ARGt4         L-Arginine Transport in via Sodium Symport   \n",
       "..          ...                                                ...   \n",
       "310   ICDCHOLt2  Isochenodeoxycholic acid transport via sodium ...   \n",
       "311   ISOCHOLt2    Isocholic acid transport via sodium cotransport   \n",
       "312  THYOCHOLt2  taurohyocholic acid transport via sodium cotra...   \n",
       "313     UCHOLt2   Ursocholic acid transport via sodium cotransport   \n",
       "314    UDCHOLt2  Ursodeoxycholic acid transport via sodium cotr...   \n",
       "\n",
       "                                                     2                      3  \n",
       "0                  adn[e] + na1[e] --> adn[c] + na1[c]       [na1[c], adn[c]]  \n",
       "1                              adprib[e] --> adprib[c]            [adprib[c]]  \n",
       "2              ala_L[e] + na1[e] --> ala_L[c] + na1[c]     [na1[c], ala_L[c]]  \n",
       "3                            aqcobal[e] --> aqcobal[c]           [aqcobal[c]]  \n",
       "4              arg_L[e] + na1[e] --> arg_L[c] + na1[c]     [na1[c], arg_L[c]]  \n",
       "..                                                 ...                    ...  \n",
       "310  icdchol[e] + 2.0 na1[e] --> icdchol[c] + 2.0 n...   [na1[c], icdchol[c]]  \n",
       "311  isochol[e] + 2.0 na1[e] --> isochol[c] + 2.0 n...   [na1[c], isochol[c]]  \n",
       "312  2.0 na1[e] + thyochol[e] --> 2.0 na1[c] + thyo...  [na1[c], thyochol[c]]  \n",
       "313    2.0 na1[e] + uchol[e] --> 2.0 na1[c] + uchol[c]     [na1[c], uchol[c]]  \n",
       "314  HC02194[e] + 2.0 na1[e] --> HC02194[c] + 2.0 n...   [na1[c], HC02194[c]]  \n",
       "\n",
       "[315 rows x 4 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if '[e]' not in str(rxns.products)\\\n",
    " if rxns.subsystem == 'Transport, extracellular'\\\n",
    " if rxns.reversibility == False]\n",
    "\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "81c4b914-cae6-471f-8053-cd14b237fa7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Alanine and aspartate metabolism',\n",
       " 'Alkaloid synthesis',\n",
       " 'Aminosugar metabolism',\n",
       " 'Androgen and estrogen synthesis and metabolism',\n",
       " 'Arachidonic acid metabolism',\n",
       " 'Arginine and proline metabolism',\n",
       " 'Beta-Alanine metabolism',\n",
       " 'Bile acid synthesis',\n",
       " 'Biotin metabolism',\n",
       " 'Blood group synthesis',\n",
       " 'Butanoate metabolism',\n",
       " 'C5-branched dibasic acid metabolism',\n",
       " 'Cholesterol metabolism',\n",
       " 'Chondroitin sulfate degradation',\n",
       " 'Chondroitin synthesis',\n",
       " 'Citric acid cycle',\n",
       " 'CoA catabolism',\n",
       " 'CoA synthesis',\n",
       " 'Cytochrome metabolism',\n",
       " 'D-alanine metabolism',\n",
       " 'Dietary fiber binding',\n",
       " 'Drug metabolism',\n",
       " 'Eicosanoid metabolism',\n",
       " 'Exchange/demand reaction',\n",
       " 'Fatty acid oxidation',\n",
       " 'Fatty acid synthesis',\n",
       " 'Folate metabolism',\n",
       " 'Fructose and mannose metabolism',\n",
       " 'Galactose metabolism',\n",
       " 'Glutamate metabolism',\n",
       " 'Glutathione metabolism',\n",
       " 'Glycerophospholipid metabolism',\n",
       " 'Glycine, serine, alanine, and threonine metabolism',\n",
       " 'Glycolysis/gluconeogenesis',\n",
       " 'Glycosphingolipid metabolism',\n",
       " 'Glyoxylate and dicarboxylate metabolism',\n",
       " 'Heme degradation',\n",
       " 'Heme synthesis',\n",
       " 'Heparan sulfate degradation',\n",
       " 'Hippurate metabolism',\n",
       " 'Histidine metabolism',\n",
       " 'Hyaluronan metabolism',\n",
       " 'Inositol phosphate metabolism',\n",
       " 'Keratan sulfate degradation',\n",
       " 'Keratan sulfate synthesis',\n",
       " 'Leukotriene metabolism',\n",
       " 'Limonene and pinene degradation',\n",
       " 'Linoleate metabolism',\n",
       " 'Lipoate metabolism',\n",
       " 'Lysine metabolism',\n",
       " 'Methionine and cysteine metabolism',\n",
       " 'Miscellaneous',\n",
       " 'N-glycan degradation',\n",
       " 'N-glycan metabolism',\n",
       " 'N-glycan synthesis',\n",
       " 'NAD metabolism',\n",
       " 'Nucleotide interconversion',\n",
       " 'Nucleotide metabolism',\n",
       " 'Nucleotide salvage pathway',\n",
       " 'Nucleotide sugar metabolism',\n",
       " 'O-glycan metabolism',\n",
       " 'Oxidative phosphorylation',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Peptide metabolism',\n",
       " 'Phenylalanine metabolism',\n",
       " 'Phosphatidylinositol phosphate metabolism',\n",
       " 'Propanoate metabolism',\n",
       " 'Protein formation',\n",
       " 'Purine catabolism',\n",
       " 'Purine synthesis',\n",
       " 'Pyrimidine catabolism',\n",
       " 'Pyrimidine synthesis',\n",
       " 'Pyruvate metabolism',\n",
       " 'R group synthesis',\n",
       " 'ROS detoxification',\n",
       " 'Sphingolipid metabolism',\n",
       " 'Squalene and cholesterol synthesis',\n",
       " 'Starch and sucrose metabolism',\n",
       " 'Steroid metabolism',\n",
       " 'Taurine and hypotaurine metabolism',\n",
       " 'Tetrahydrobiopterin metabolism',\n",
       " 'Thiamine metabolism',\n",
       " 'Transport, endoplasmic reticular',\n",
       " 'Transport, extracellular',\n",
       " 'Transport, golgi apparatus',\n",
       " 'Transport, lysosomal',\n",
       " 'Transport, mitochondrial',\n",
       " 'Transport, nuclear',\n",
       " 'Transport, peroxisomal',\n",
       " 'Triacylglycerol synthesis',\n",
       " 'Tryptophan metabolism',\n",
       " 'Tyrosine metabolism',\n",
       " 'Ubiquinone synthesis',\n",
       " 'Urea cycle',\n",
       " 'Valine, leucine, and isoleucine metabolism',\n",
       " 'Vitamin A metabolism',\n",
       " 'Vitamin B12 metabolism',\n",
       " 'Vitamin B2 metabolism',\n",
       " 'Vitamin B6 metabolism',\n",
       " 'Vitamin C metabolism',\n",
       " 'Vitamin D metabolism',\n",
       " 'Vitamin E metabolism',\n",
       " 'Xenobiotics metabolism'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set([rxns.subsystem for rxns in mod_recon3d.reactions])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2db35190-bd07-4435-8d39-9fa89f393663",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DM_13_cis_oretn_n_</td>\n",
       "      <td>13-Cis-Oxo-Retinoic Acid Demand</td>\n",
       "      <td>13_cis_oretn[n] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DM_13_cis_retn_n_</td>\n",
       "      <td>13-Cis-Retinoate Demand</td>\n",
       "      <td>13_cis_retn[n] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DM_Asn_X_Ser_Thr_ly_</td>\n",
       "      <td>Demand for Asn-X-Ser/Thr, Lysosomal</td>\n",
       "      <td>Asn_X_Ser_Thr[l] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DM_avite2_c_</td>\n",
       "      <td>Demand for Alpha-Tocotrienol</td>\n",
       "      <td>avite2[c] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DM_bvite_c_</td>\n",
       "      <td>Demand for Beta-Tocopherol</td>\n",
       "      <td>bvite[c] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1782</th>\n",
       "      <td>DM_Lkynr[c]</td>\n",
       "      <td>DM_Lkynr[c]</td>\n",
       "      <td>Lkynr[c] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1783</th>\n",
       "      <td>DM_cbl2[m]</td>\n",
       "      <td>DM_cbl2[m]</td>\n",
       "      <td>cbl2[m] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1784</th>\n",
       "      <td>DM_1a2425thvitd2[m]</td>\n",
       "      <td>DM_1a2425thvitd2[m]</td>\n",
       "      <td>1a2425thvitd2[m] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1785</th>\n",
       "      <td>DM_btn[m]</td>\n",
       "      <td>DM_btn[m]</td>\n",
       "      <td>btn[m] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786</th>\n",
       "      <td>DM_btn[n]</td>\n",
       "      <td>DM_btn[n]</td>\n",
       "      <td>btn[n] --&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1787 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         0                                    1  \\\n",
       "0       DM_13_cis_oretn_n_      13-Cis-Oxo-Retinoic Acid Demand   \n",
       "1        DM_13_cis_retn_n_              13-Cis-Retinoate Demand   \n",
       "2     DM_Asn_X_Ser_Thr_ly_  Demand for Asn-X-Ser/Thr, Lysosomal   \n",
       "3             DM_avite2_c_         Demand for Alpha-Tocotrienol   \n",
       "4              DM_bvite_c_           Demand for Beta-Tocopherol   \n",
       "...                    ...                                  ...   \n",
       "1782           DM_Lkynr[c]                          DM_Lkynr[c]   \n",
       "1783            DM_cbl2[m]                           DM_cbl2[m]   \n",
       "1784   DM_1a2425thvitd2[m]                  DM_1a2425thvitd2[m]   \n",
       "1785             DM_btn[m]                            DM_btn[m]   \n",
       "1786             DM_btn[n]                            DM_btn[n]   \n",
       "\n",
       "                          2   3  \n",
       "0      13_cis_oretn[n] -->   []  \n",
       "1       13_cis_retn[n] -->   []  \n",
       "2     Asn_X_Ser_Thr[l] -->   []  \n",
       "3            avite2[c] -->   []  \n",
       "4             bvite[c] -->   []  \n",
       "...                     ...  ..  \n",
       "1782          Lkynr[c] -->   []  \n",
       "1783           cbl2[m] -->   []  \n",
       "1784  1a2425thvitd2[m] -->   []  \n",
       "1785            btn[m] -->   []  \n",
       "1786            btn[n] -->   []  \n",
       "\n",
       "[1787 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if rxns.subsystem == 'Exchange/demand reaction']\n",
    "\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "03237a96-85f6-43a5-bf8e-27d2da0bde93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mod_recon3d.reactions.get_by_id('DM_13_cis_oretn_n_').products)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3e7a6a41-adf7-44cd-a9ed-b67f5b93f171",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>DM_13_cis_oretn_n_</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>13-Cis-Oxo-Retinoic Acid Demand</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x1c1870801f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>13_cis_oretn[n] --></p>\n",
       "                    <p style='text-align:right'>13-Cis-Oxo-Retinoic Acid --></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction DM_13_cis_oretn_n_ at 0x1c1870801f0>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(mod_recon3d.reactions.get_by_id('DM_13_cis_oretn_n_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1dcca5cb-0470-41e2-9f17-92c26683d58a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EX_lthstrl[e]</td>\n",
       "      <td>Exchange of Lathosterol</td>\n",
       "      <td>lthstrl[e] &lt;=&gt;</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               0                        1                2   3\n",
       "0  EX_lthstrl[e]  Exchange of Lathosterol  lthstrl[e] <=>   []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction, rxns.products] for rxns in mod_recon3d.reactions \\\n",
    " if len(rxns.products) == 0\\\n",
    " or len(rxns.reactants) == 0\\\n",
    " if rxns.subsystem == 'Transport, extracellular']\n",
    "\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)\n",
    "pd.DataFrame(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d3b2c804-7e71-45f7-a5c7-9bb5bc513838",
   "metadata": {},
   "outputs": [],
   "source": [
    "l = [[rxns.id, rxns.name, rxns.reaction] for rxns in mod_recon3d.reactions \\\n",
    " if '[e]' not in str(rxns.reaction)\\\n",
    " if rxns.subsystem == 'Transport, extracellular']\n",
    "\n",
    "pd.DataFrame(l).to_clipboard(excel=True, sep=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "cef53091-c1d8-4b16-9b19-718a86bbdbf9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'cgly[e] + 2.0 h[e] --> cgly[c] + 2.0 h[c]'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_recon3d.reactions.get_by_id('CGLYt3_2_').reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a06db048-d32d-478e-adc0-8bd4d81d16bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__add__',\n",
       " '__class__',\n",
       " '__copy__',\n",
       " '__deepcopy__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__iadd__',\n",
       " '__imul__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__isub__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__mul__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__radd__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__sub__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_annotation',\n",
       " '_associate_gene',\n",
       " '_check_bounds',\n",
       " '_dissociate_gene',\n",
       " '_genes',\n",
       " '_gpr',\n",
       " '_id',\n",
       " '_lower_bound',\n",
       " '_metabolites',\n",
       " '_model',\n",
       " '_repr_html_',\n",
       " '_set_id_with_model',\n",
       " '_update_awareness',\n",
       " '_upper_bound',\n",
       " 'add_metabolites',\n",
       " 'annotation',\n",
       " 'boundary',\n",
       " 'bounds',\n",
       " 'build_reaction_from_string',\n",
       " 'build_reaction_string',\n",
       " 'check_mass_balance',\n",
       " 'compartments',\n",
       " 'copy',\n",
       " 'delete',\n",
       " 'flux',\n",
       " 'flux_expression',\n",
       " 'forward_variable',\n",
       " 'functional',\n",
       " 'gene_name_reaction_rule',\n",
       " 'gene_reaction_rule',\n",
       " 'genes',\n",
       " 'get_coefficient',\n",
       " 'get_coefficients',\n",
       " 'get_compartments',\n",
       " 'gpr',\n",
       " 'id',\n",
       " 'knock_out',\n",
       " 'lower_bound',\n",
       " 'metabolites',\n",
       " 'model',\n",
       " 'name',\n",
       " 'notes',\n",
       " 'objective_coefficient',\n",
       " 'products',\n",
       " 'reactants',\n",
       " 'reaction',\n",
       " 'reduced_cost',\n",
       " 'remove_from_model',\n",
       " 'reverse_id',\n",
       " 'reverse_variable',\n",
       " 'reversibility',\n",
       " 'subsystem',\n",
       " 'subtract_metabolites',\n",
       " 'summary',\n",
       " 'update_genes_from_gpr',\n",
       " 'update_variable_bounds',\n",
       " 'upper_bound',\n",
       " 'x',\n",
       " 'y']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(mod_recon3d.reactions.get_by_id('CGLYt3_2_'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6f8be247-8385-4440-8c0f-c4c55d2af0f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Carnitine shuttle (peroxisomal)',\n",
       " 'Beta oxidation of odd-chain fatty acids (peroxisomal)',\n",
       " 'Artificial reactions',\n",
       " 'Nicotinate and nicotinamide metabolism',\n",
       " 'Cysteine and methionine metabolism',\n",
       " 'Ether lipid metabolism',\n",
       " 'Beta oxidation of branched-chain fatty acids (mitochondrial)',\n",
       " 'Riboflavin metabolism',\n",
       " 'Beta oxidation of odd-chain fatty acids (mitochondrial)',\n",
       " 'Cholesterol biosynthesis 2',\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis',\n",
       " 'Amino sugar and nucleotide sugar metabolism',\n",
       " 'Beta-alanine metabolism',\n",
       " 'Fatty acid biosynthesis (odd-chain)',\n",
       " 'Biopterin metabolism',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (peroxisomal)',\n",
       " 'Acyl-CoA hydrolysis',\n",
       " 'Fatty acid elongation (even-chain)',\n",
       " 'Serotonin and melatonin biosynthesis',\n",
       " 'Glycine, serine and threonine metabolism',\n",
       " 'Beta oxidation of phytanic acid (peroxisomal)',\n",
       " 'Purine metabolism',\n",
       " 'Androgen metabolism',\n",
       " 'Glycerolipid metabolism',\n",
       " 'Protein assembly',\n",
       " 'Blood group biosynthesis',\n",
       " 'Fatty acid elongation (odd-chain)',\n",
       " 'Fatty acid biosynthesis (unsaturated)',\n",
       " 'Pentose and glucuronate interconversions',\n",
       " 'Isolated',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (peroxisomal)',\n",
       " 'Pool reactions',\n",
       " 'Aminoacyl-tRNA biosynthesis',\n",
       " 'Carnitine shuttle (mitochondrial)',\n",
       " 'Retinol metabolism',\n",
       " 'Cholesterol biosynthesis 3 (Kandustch-Russell pathway)',\n",
       " 'Glycosylphosphatidylinositol (GPI)-anchor biosynthesis',\n",
       " 'Beta oxidation of even-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (mitochondrial)',\n",
       " 'Alkaloids biosynthesis',\n",
       " 'Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism',\n",
       " 'Carnitine shuttle (endoplasmic reticular)',\n",
       " 'Fatty acid biosynthesis (even-chain)',\n",
       " 'Glycolysis / Gluconeogenesis',\n",
       " 'Bile acid recycling',\n",
       " 'Exchange/demand reactions',\n",
       " 'Omega-3 fatty acid metabolism',\n",
       " 'Acylglycerides metabolism',\n",
       " 'Glycosphingolipid biosynthesis-ganglio series',\n",
       " 'Estrogen metabolism',\n",
       " 'Bile acid biosynthesis',\n",
       " 'Keratan sulfate biosynthesis',\n",
       " 'Fatty acid desaturation (even-chain)',\n",
       " 'Pantothenate and CoA biosynthesis',\n",
       " 'Glucocorticoid biosynthesis',\n",
       " 'Beta oxidation of even-chain fatty acids (peroxisomal)',\n",
       " 'Sulfur metabolism',\n",
       " 'Protein degradation',\n",
       " 'Alanine, aspartate and glutamate metabolism',\n",
       " 'Porphyrin metabolism',\n",
       " 'Glycosphingolipid biosynthesis-globo series',\n",
       " 'Chondroitin / heparan sulfate biosynthesis',\n",
       " 'Terpenoid backbone biosynthesis',\n",
       " 'Omega-6 fatty acid metabolism',\n",
       " 'Transport reactions',\n",
       " 'Cholesterol biosynthesis 1 (Bloch pathway)',\n",
       " 'Pyrimidine metabolism',\n",
       " 'Formation and hydrolysis of cholesterol esters',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (peroxisomal)',\n",
       " 'Carnitine shuttle (cytosolic)',\n",
       " 'Fatty acid activation (cytosolic)',\n",
       " 'Fatty acid activation (endoplasmic reticular)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (mitochondrial)',\n",
       " 'Protein modification',\n",
       " 'Fatty acid desaturation (odd-chain)',\n",
       " 'Metabolism of other amino acids',\n",
       " 'Glycosphingolipid biosynthesis-lacto and neolacto series',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Prostaglandin biosynthesis',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (mitochondrial)',\n",
       " 'Beta oxidation of poly-unsaturated fatty acids (mitochondrial)',\n",
       " 'Ascorbate and aldarate metabolism',\n",
       " 'Fatty acid biosynthesis']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set([rxns.subsystem for rxns in mod_sbml.reactions]) - set([rxns.subsystem for rxns in mod_recon3d.reactions]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "24d31029-dad6-4992-9287-212eb26a9e95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Acyl-CoA hydrolysis',\n",
       " 'Acylglycerides metabolism',\n",
       " 'Alanine, aspartate and glutamate metabolism',\n",
       " 'Alkaloids biosynthesis',\n",
       " 'Amino sugar and nucleotide sugar metabolism',\n",
       " 'Aminoacyl-tRNA biosynthesis',\n",
       " 'Androgen metabolism',\n",
       " 'Arachidonic acid metabolism',\n",
       " 'Arginine and proline metabolism',\n",
       " 'Artificial reactions',\n",
       " 'Ascorbate and aldarate metabolism',\n",
       " 'Beta oxidation of branched-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (mitochondrial)',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (peroxisomal)',\n",
       " 'Beta oxidation of even-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of even-chain fatty acids (peroxisomal)',\n",
       " 'Beta oxidation of odd-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of odd-chain fatty acids (peroxisomal)',\n",
       " 'Beta oxidation of phytanic acid (peroxisomal)',\n",
       " 'Beta oxidation of poly-unsaturated fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (peroxisomal)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (peroxisomal)',\n",
       " 'Beta-alanine metabolism',\n",
       " 'Bile acid biosynthesis',\n",
       " 'Bile acid recycling',\n",
       " 'Biopterin metabolism',\n",
       " 'Biotin metabolism',\n",
       " 'Blood group biosynthesis',\n",
       " 'Butanoate metabolism',\n",
       " 'C5-branched dibasic acid metabolism',\n",
       " 'Carnitine shuttle (cytosolic)',\n",
       " 'Carnitine shuttle (endoplasmic reticular)',\n",
       " 'Carnitine shuttle (mitochondrial)',\n",
       " 'Carnitine shuttle (peroxisomal)',\n",
       " 'Cholesterol biosynthesis 1 (Bloch pathway)',\n",
       " 'Cholesterol biosynthesis 2',\n",
       " 'Cholesterol biosynthesis 3 (Kandustch-Russell pathway)',\n",
       " 'Cholesterol metabolism',\n",
       " 'Chondroitin / heparan sulfate biosynthesis',\n",
       " 'Chondroitin sulfate degradation',\n",
       " 'CoA catabolism',\n",
       " 'CoA synthesis',\n",
       " 'Cysteine and methionine metabolism',\n",
       " 'Dietary fiber binding',\n",
       " 'Drug metabolism',\n",
       " 'Eicosanoid metabolism',\n",
       " 'Estrogen metabolism',\n",
       " 'Ether lipid metabolism',\n",
       " 'Exchange/demand reactions',\n",
       " 'Fatty acid activation (cytosolic)',\n",
       " 'Fatty acid activation (endoplasmic reticular)',\n",
       " 'Fatty acid biosynthesis',\n",
       " 'Fatty acid biosynthesis (even-chain)',\n",
       " 'Fatty acid biosynthesis (odd-chain)',\n",
       " 'Fatty acid biosynthesis (unsaturated)',\n",
       " 'Fatty acid desaturation (even-chain)',\n",
       " 'Fatty acid desaturation (odd-chain)',\n",
       " 'Fatty acid elongation (even-chain)',\n",
       " 'Fatty acid elongation (odd-chain)',\n",
       " 'Fatty acid oxidation',\n",
       " 'Folate metabolism',\n",
       " 'Formation and hydrolysis of cholesterol esters',\n",
       " 'Fructose and mannose metabolism',\n",
       " 'Galactose metabolism',\n",
       " 'Glucocorticoid biosynthesis',\n",
       " 'Glutathione metabolism',\n",
       " 'Glycerolipid metabolism',\n",
       " 'Glycerophospholipid metabolism',\n",
       " 'Glycine, serine and threonine metabolism',\n",
       " 'Glycolysis / Gluconeogenesis',\n",
       " 'Glycosphingolipid biosynthesis-ganglio series',\n",
       " 'Glycosphingolipid biosynthesis-globo series',\n",
       " 'Glycosphingolipid biosynthesis-lacto and neolacto series',\n",
       " 'Glycosphingolipid metabolism',\n",
       " 'Glycosylphosphatidylinositol (GPI)-anchor biosynthesis',\n",
       " 'Heme degradation',\n",
       " 'Heme synthesis',\n",
       " 'Heparan sulfate degradation',\n",
       " 'Hippurate metabolism',\n",
       " 'Histidine metabolism',\n",
       " 'Inositol phosphate metabolism',\n",
       " 'Isolated',\n",
       " 'Keratan sulfate biosynthesis',\n",
       " 'Keratan sulfate degradation',\n",
       " 'Leukotriene metabolism',\n",
       " 'Linoleate metabolism',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Lysine metabolism',\n",
       " 'Metabolism of other amino acids',\n",
       " 'Miscellaneous',\n",
       " 'N-glycan metabolism',\n",
       " 'Nicotinate and nicotinamide metabolism',\n",
       " 'Nucleotide metabolism',\n",
       " 'O-glycan metabolism',\n",
       " 'Omega-3 fatty acid metabolism',\n",
       " 'Omega-6 fatty acid metabolism',\n",
       " 'Oxidative phosphorylation',\n",
       " 'Pantothenate and CoA biosynthesis',\n",
       " 'Pentose and glucuronate interconversions',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Peptide metabolism',\n",
       " 'Phenylalanine metabolism',\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis',\n",
       " 'Phosphatidylinositol phosphate metabolism',\n",
       " 'Pool reactions',\n",
       " 'Porphyrin metabolism',\n",
       " 'Propanoate metabolism',\n",
       " 'Prostaglandin biosynthesis',\n",
       " 'Protein assembly',\n",
       " 'Protein degradation',\n",
       " 'Protein modification',\n",
       " 'Purine metabolism',\n",
       " 'Pyrimidine metabolism',\n",
       " 'Pyruvate metabolism',\n",
       " 'ROS detoxification',\n",
       " 'Retinol metabolism',\n",
       " 'Riboflavin metabolism',\n",
       " 'Serotonin and melatonin biosynthesis',\n",
       " 'Sphingolipid metabolism',\n",
       " 'Starch and sucrose metabolism',\n",
       " 'Steroid metabolism',\n",
       " 'Sulfur metabolism',\n",
       " 'Terpenoid backbone biosynthesis',\n",
       " 'Thiamine metabolism',\n",
       " 'Transport reactions',\n",
       " 'Triacylglycerol synthesis',\n",
       " 'Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism',\n",
       " 'Tryptophan metabolism',\n",
       " 'Tyrosine metabolism',\n",
       " 'Ubiquinone synthesis',\n",
       " 'Urea cycle',\n",
       " 'Valine, leucine, and isoleucine metabolism',\n",
       " 'Vitamin A metabolism',\n",
       " 'Vitamin B12 metabolism',\n",
       " 'Vitamin B2 metabolism',\n",
       " 'Vitamin B6 metabolism',\n",
       " 'Vitamin C metabolism',\n",
       " 'Vitamin D metabolism',\n",
       " 'Vitamin E metabolism',\n",
       " 'Xenobiotics metabolism'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set([rxns.subsystem for rxns in mod_sbml.reactions])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d6d1a5ef-4da5-44ba-a80c-dce770013223",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction MAR13077 at 0x1c11494b010>]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[rxn for rxn in mod_sbml.reactions if 'CFTR' in rxn.name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2c1970e4-6351-4b0b-876e-fde142c39701",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP:acetate adenylyltransferase',\n",
       " 'ATP:D-fructose-6-phosphate 1-phosphotransferase',\n",
       " 'ATP:pyruvate 2-O-phosphotransferase',\n",
       " 'ATP:3-phospho-D-glycerate 1-phosphotransferase',\n",
       " 'ATP:D-glucose 6-phosphotransferase',\n",
       " 'ATP:alpha-D-galactose 1-phosphotransferase',\n",
       " 'ATP:D-tagatose 6-phosphotransferase',\n",
       " 'ATP:D-glyceraldehyde 3-phosphotransferase',\n",
       " 'ATP:D-fructose 1-phosphotransferase',\n",
       " 'ATP:D-fructose 6-phosphotransferase',\n",
       " 'ATP:6-deoxy-L-galactose 1-phosphotransferase',\n",
       " 'ATP:D-mannose 6-phosphotransferase',\n",
       " 'ATP:D-xylulose 5-phosphotransferase',\n",
       " 'ATP:D-ribulose 5-phosphotransferase',\n",
       " 'ATP:propanoate adenylyltransferase',\n",
       " 'ATP:propanoate adenylyltransferase',\n",
       " 'ATP:D-ribose-5-phosphate diphosphotransferase',\n",
       " 'ATP:D-ribose 5-phosphotransferase',\n",
       " 'ATP:D-gluconate 6-phosphotransferase',\n",
       " 'ATP:Sedoheptulose 7-phosphate 1-phosphotransferase',\n",
       " 'ATP:sedoheptulose 7-phosphate',\n",
       " 'ATP:sedoheptulose 1-phosphotransferase',\n",
       " 'ATP:2-deoxy-D-ribose 5-phosphotransferase',\n",
       " 'ATP diphosphate-lyase (cyclizing',\n",
       " 'ATP:AMP phosphotransferase',\n",
       " 'ATP:AMP phosphotransferase',\n",
       " \"ATP:adenosine 5'-phosphotransferase\",\n",
       " 'ATP:GMP phosphotransferase',\n",
       " 'ATP:GMP phosphotransferase',\n",
       " 'dATP:pyruvate 2-O-phosphotransferase',\n",
       " \"ATP:guanosine 5'-phosphotransferase\",\n",
       " 'ATP:dAMP phosphotransferase',\n",
       " 'dATP:D-glucose 6-phosphotransferase',\n",
       " 'dATP:D-fructose 6-phosphotransferase',\n",
       " 'dATP:D-mannose 6-phosphotransferase',\n",
       " 'dATP:D-glucosamine 6-phosphotransferase',\n",
       " \"ATP:deoxyguanosine 5'-phosphotransferase\",\n",
       " 'ATP:dGMP phosphotransferase',\n",
       " 'ATP:GDP phosphotransferase',\n",
       " 'ATP:GDP phosphotransferase',\n",
       " 'ATP:dADP phosphotransferase',\n",
       " 'ATP:IDP phosphotransferase',\n",
       " 'ATP:dADP phosphotransferase',\n",
       " 'ATP:dGDP phosphotransferase',\n",
       " 'ATP:dGDP phosphotransferase',\n",
       " 'ATP:UDP phosphotransferase',\n",
       " 'ATP:CDP phosphotransferase',\n",
       " 'ATP:CDP phosphotransferase',\n",
       " 'ATP:dCDP phosphotransferase',\n",
       " 'ATP:dUDP phosphotransferase',\n",
       " 'ATP:dUDP phosphotransferase',\n",
       " 'ATP:dTDP phosphotransferase',\n",
       " 'ATP:UMP phosphotransferase',\n",
       " 'ATP:dCMP phosphotransferase',\n",
       " \"ATP:deoxycitidine 5'-phosphotransferase\",\n",
       " 'ATP:dTMP phosphotransferase',\n",
       " \"ATP:uridine 5'-phosphotransferase\",\n",
       " 'ATP diphosphohydrolase (diphosphate-forming)',\n",
       " 'ATP:dUMP phosphotransferase',\n",
       " 'ATP:CMP phosphotransferase',\n",
       " 'ATP:CMP phosphotransferase',\n",
       " \"ATP:thymidine 5'-phosphotransferase\",\n",
       " \"ATP:adenosine 5'-phosphotransferase\",\n",
       " \"ATP:cytidine 5'-phosphotransferase\",\n",
       " 'ATP:sulfate adenylyltransferase',\n",
       " \"ATP:adenylylsulfate 3'-phosphotransferase\",\n",
       " \"ATP:thymidine 5'-phosphotransferase\",\n",
       " \"ATP:deoxyguanosine 5'-phosphotransferase\",\n",
       " \"ATP:deoxyadenosine 5'-phosphotransferase\",\n",
       " 'ATP:dUMP phosphotransferase',\n",
       " \"ATP:deoxyuridine 5'-phosphotransferase\",\n",
       " \"ATP:deoxyuridine 5'-phosphotransferase\",\n",
       " 'ATP:UMP phosphotransferase',\n",
       " 'ATP:UMP phosphotransferase',\n",
       " 'ATP:dIDP phosphotransferase',\n",
       " 'ATP:dIDP phosphotransferase',\n",
       " 'ATP:GDP phosphotransferase',\n",
       " 'ATP:UDP phosphotransferase',\n",
       " 'ATP:UDP phosphotransferase',\n",
       " 'ATP:CDP phosphotransferase',\n",
       " 'ATP:dTDP phosphotransferase',\n",
       " 'ATP:dTDP phosphotransferase',\n",
       " 'ATP:dGDP phosphotransferase',\n",
       " 'ATP:dUDP phosphotransferase',\n",
       " 'ATP:dCDP phosphotransferase',\n",
       " 'ATP:dCDP phosphotransferase',\n",
       " 'ATP:dADP phosphotransferase',\n",
       " 'ATP:IDP phosphotransferase',\n",
       " 'ATP:dIDP phosphotransferase',\n",
       " 'ATP:IDP phosphotransferase',\n",
       " \"ATP:cytidine 5'-phosphotransferase\",\n",
       " 'ATP:CMP phosphotransferase',\n",
       " 'ATP:dCMP phosphotransferase',\n",
       " 'ATP diphosphohydrolase (phosphate-forming)',\n",
       " \"ATP:deoxycitidine 5'-phosphotransferase\",\n",
       " \"ATP:deoxycitidine 5'-phosphotransferase\",\n",
       " \"ATP:inosine 5'-phosphotransferase\",\n",
       " 'ATP:L-glutamate 5-phosphotransferase',\n",
       " 'ATP:(R)-glycerate 3-phosphotransferase',\n",
       " 'ATP:(R)-glycerate 3-phosphotransferase',\n",
       " 'ATP:L-homoserine O-phosphotransferase',\n",
       " 'ATP:L-methionine S-adenosyltransferase',\n",
       " '5-oxo-L-proline amidohydrolase (ATP-hydrolysing)',\n",
       " 'ATP:selenate adenylyltransferase',\n",
       " \"ATP:adenylylsulfate 3'-phosphotransferase\",\n",
       " 'ATP:selenide, water phosphotransferase',\n",
       " 'ATP:L-selenomethione S-adenosyltransferase',\n",
       " 'ATP:D-glucosamine 6-phosphotransferase',\n",
       " 'ATP:N-acetyl-D-glucosamine 6-phosphotransferase',\n",
       " 'ATP:N-acyl-D-mannosamine 6-phosphotransferase',\n",
       " 'ATP:(R)-glycerate 2-phosphotransferase',\n",
       " 'ATP phosphohydrolase',\n",
       " 'ATP:1,2-diacylglycerol 3-phosphotransferase',\n",
       " 'ATP:(R)-mevalonate 5-phosphotransferase',\n",
       " 'ATP:(R)-5-phosphomevalonate phosphotransferase',\n",
       " 'ATP:(R)-5-diphosphomevalonate carboxy-lyase (adding ATP',\n",
       " 'ATP:sphinganine 1-phosphotransferase',\n",
       " 'ATP:ceramide 1-phosphotransferase',\n",
       " 'ATP:sphingosine 1-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:acylglycerol 3-phosphotransferase',\n",
       " 'ATP:glycerol 3-phosphotransferase',\n",
       " 'ATP:glycerol 3-phosphotransferase',\n",
       " 'ATP:choline phosphotransferase',\n",
       " 'ATP:ethanolamine O-phosphotransferase',\n",
       " 'ATP:glycerone phosphotransferase',\n",
       " 'ATP:beta-D-ribosylnicotinate 5-phosphotransferase',\n",
       " 'ATP:nicotinamide-nucleotide adenylyltransferase',\n",
       " \"ATP:NAD+ 2'-phosphotransferase\",\n",
       " \"ATP:NAD+ 2'-phosphotransferase\",\n",
       " 'ATP:nicotinamide-nucleotide adenylyltransferase',\n",
       " \"ATP:NAD+ 2'-phosphotransferase\",\n",
       " \"ATP:dephospho-CoA 3'-phosphotransferase\",\n",
       " \"ATP:dephospho-CoA 3'-phosphotransferase\",\n",
       " \"ATP:pantothenate 4'-phosphotransferase\",\n",
       " \"ATP:pantothenate 4'-phosphotransferase\",\n",
       " \"ATP:pantothenate 4'-phosphotransferase\",\n",
       " \"ATP:pantetheine-4'-phosphate adenylyltransferase\",\n",
       " \"ATP:pantothenate 4'-phosphotransferase\",\n",
       " \"ATP:pantothenate 4'-phosphotransferase\",\n",
       " \"ATP:pantetheine-4'-phosphate adenylyltransferase\",\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol 3-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-3-phosphate 5-phosphotransferase',\n",
       " 'ATP:1-Phosphatidyl-1D-myo-inositol 4-phosphotransferase',\n",
       " 'ATP:1-Phosphatidyl-1D-myo-inositol 4-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4-phosphate 3-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4-phosphate 5-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-5-phosphate 4-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4,5-bisphosphate 3-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5-trisphosphate 3-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,3,4-trisphosphate 5-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,3,4-trisphosphate 6-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,3,4,6-tetrakisphosphate 5-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5-trisphosphate 6-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5,6-tetrakisphosphate 3-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-3,4,5,6-tetrakisphosphate 1-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol 1,3,4,5,6-pentakisphosphate 2-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-hexakisphosphate 5-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-5-diphosphate-pentakisphosphate phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,3,4,6-tetrakisphosphate 5-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5,6-tetrakisphosphate 3-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5-trisphosphate 6-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,4,5-trisphosphate 3-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4,5-bisphosphate 3-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4-phosphate 3-phosphotransferase',\n",
       " 'ATP:1-Phosphatidyl-1D-myo-inositol 4-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-4-phosphate 5-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-5-phosphate 4-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol 3-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol 3-phosphotransferase',\n",
       " 'ATP:1-Phosphatidyl-1D-myo-inositol 4-phosphotransferase',\n",
       " 'ATP:molybdopterin adenylyltransferase',\n",
       " \"ATP:riboflavin 5'-phosphotransferase\",\n",
       " 'ATP:FMN adenylyltransferase',\n",
       " 'ATP:thiamine diphosphotransferase',\n",
       " 'ATP:thiamine-diphosphate phosphotransferase',\n",
       " 'ATP:cob(I)alamin Co-beta-adenosyltransferase',\n",
       " 'ATP:cob(I)alamin Co-beta-adenosyltransferase',\n",
       " \"ATP:pyridoxal 5'-phosphotransferase\",\n",
       " \"ATP:pyridoxine 5'-phosphotransferase\",\n",
       " \"ATP:pyridoxal 5'-phosphotransferase\",\n",
       " 'ATP:phosphorylase-b phosphotransferase',\n",
       " 'ATP:myosin-light-chain O-phosphotransferase',\n",
       " 'ATP:rhodopsin phosphotransferase',\n",
       " 'ATP:[pyruvate dehydrogenase (lipoamide)] phosphotransferase',\n",
       " 'ATP:beta-adrenergic-receptor phosphotransferase',\n",
       " 'ATP:[3-methyl-2-oxobutanoate dehydrogenase (lipoamide)] phosphotransferase',\n",
       " 'ATP phosphohydrolase',\n",
       " 'ATP:ADP phosphatransferase',\n",
       " 'ATP:ADP phosphatransferase',\n",
       " 'ATP:ADP phosphatransferase',\n",
       " 'ATP:ADP phosphatransferase',\n",
       " '5-Hydroxy-L-Tryptophan Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Acetylcholine Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Adrenaline Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'V-Type ATPase, H+ Transporting, Lysosomal',\n",
       " 'B-Alanine Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Ca ATPase',\n",
       " 'Secretion of Dopamine via Secretory Vesicle (ATP Driven)',\n",
       " 'Exchange of ATP',\n",
       " 'GABA Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Secretion of Glycine  via Secretory Vesicle (ATP Driven)',\n",
       " 'Histamine Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'H+/K+ Gastric/Non-Gastric P-ATPase And ABC ATPase',\n",
       " 'Metanephrine Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Mevalonate Kinase (ATP)',\n",
       " 'Secretion of Noradrenaline  via Secretory Vesicle (ATP Driven)',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.204.5',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.208.2',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.211.1',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.211.1',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.203.3',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.203.3',\n",
       " 'Transport of dATP into Mitochondria',\n",
       " 'Arachidonate (N-C18:1) Transport by FATP',\n",
       " 'Folate Transport by OATP',\n",
       " 'Lignocerate Transport by FATP',\n",
       " 'Transport of 12-Hydroperoxyeicosa-5, 8, 10, 14-Tetraenoate, by FATP',\n",
       " 'Transport of 11, 12-EET, FATP',\n",
       " 'Transport of 12 (13)-Epome, FATP',\n",
       " 'Transport of 12S-HHT, FATP',\n",
       " 'Transport of Dodecanedioic Acid by FATP',\n",
       " 'Transport of Hexadecanediocacid by FATP',\n",
       " '3-Methoxytyramine Secretion via Secretory Vesicle (ATP Driven)',\n",
       " 'L-Normetanephrine Seceretion via Secretory Vesicle (ATP Driven)',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport by OATP1B Carriers into Extracellular Space',\n",
       " 'Transport by OATP1B Carriers into Extracellular Space',\n",
       " 'Export Mediated by ATP Binding Cassette (ABC) Transporters',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'Transport (ATP-Dependent) into Extracellular Space',\n",
       " 'ADP/ATP Transporter, Golgi']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[rxn.name for rxn in mod_sbml.reactions if 'ATP' in rxn.name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "722f2677-812a-4910-9b64-369ace2ca2de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['(S)-Lactate:NAD+ oxidoreductase',\n",
       " '(S)-Lactate:NAD+ oxidoreductase',\n",
       " '(R)-Lactate:ferricytochrome-c 2-oxidoreductase',\n",
       " '(R)-Lactate:NAD+ oxidoreductase',\n",
       " '(S)-Lactate:ferricytochrome-c 2-oxidoreductase',\n",
       " '(S)-Lactate:NAD+ oxidoreductase',\n",
       " '3- (4-Hydroxyphenyl-)Lactate Formation',\n",
       " 'Transport of L-Lactate via Diffusion, Mitochondrial',\n",
       " 'Lactate Transport',\n",
       " 'Formation of Phenyl-Lactate',\n",
       " 'Transport of Phenyl Lactate',\n",
       " 'Transport of 15-Keto-Prostaglandin F2A, Lactate Antiport',\n",
       " 'Transport of 3- (4-Hydroxyphenyl)Lactate',\n",
       " 'Transport of Prostaglandin A2, Lactate Antiport',\n",
       " 'Transport of Prostaglandin J2, Lactate Antiport',\n",
       " 'Transport of Prostaglandin E3, Lactate Antiport',\n",
       " 'Transport of 6-Oxo-Prostaglandin F1Alpha, Lactate Antiport',\n",
       " 'Transport of 15-Deoxy-Pgd2, Antiport with Lactate',\n",
       " 'Exchange of 3- (4-Hydroxyphenyl)Lactate',\n",
       " 'Exchange of Phenyl Lactate',\n",
       " 'Transport of 3-(4-Hydroxyphenyl)Lactate, Mitochondrial',\n",
       " 'Formation of Vanil-Lactate',\n",
       " 'Secretion of Vanil-Lactate',\n",
       " 'Exchange of Vanil-Lactate']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[rxn.name for rxn in mod_sbml.reactions if 'Lactate' in rxn.name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "5691f54f-c273-42c3-a588-af8f398879aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>MAR07116</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x1c1126d5ab0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>MAM01403e <=></p>\n",
       "                    <p style='text-align:right'>bromobenzene <=></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1000.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction MAR07116 at 0x1c1126d5ab0>"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_sbml.exchanges[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8f1818f8-7b7c-42d7-86c9-42511a9c007a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__class__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__enter__',\n",
       " '__eq__',\n",
       " '__exit__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_annotation',\n",
       " '_compartments',\n",
       " '_contexts',\n",
       " '_id',\n",
       " '_populate_solver',\n",
       " '_repr_html_',\n",
       " '_sbml',\n",
       " '_set_id_with_model',\n",
       " '_solver',\n",
       " '_tolerance',\n",
       " 'add_boundary',\n",
       " 'add_cons_vars',\n",
       " 'add_groups',\n",
       " 'add_metabolites',\n",
       " 'add_reactions',\n",
       " 'annotation',\n",
       " 'boundary',\n",
       " 'compartments',\n",
       " 'constraints',\n",
       " 'copy',\n",
       " 'demands',\n",
       " 'exchanges',\n",
       " 'genes',\n",
       " 'get_associated_groups',\n",
       " 'groups',\n",
       " 'id',\n",
       " 'medium',\n",
       " 'merge',\n",
       " 'metabolites',\n",
       " 'name',\n",
       " 'notes',\n",
       " 'objective',\n",
       " 'objective_direction',\n",
       " 'optimize',\n",
       " 'problem',\n",
       " 'reactions',\n",
       " 'remove_cons_vars',\n",
       " 'remove_groups',\n",
       " 'remove_metabolites',\n",
       " 'remove_reactions',\n",
       " 'repair',\n",
       " 'sinks',\n",
       " 'slim_optimize',\n",
       " 'solver',\n",
       " 'summary',\n",
       " 'tolerance',\n",
       " 'variables']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(mod_sbml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b1e5e66d-649b-47f8-b5bd-3ad1232c5e58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_sbml.reactions[0].boundary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5023d262-1206-4575-a95b-af695a4a35d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__add__',\n",
       " '__class__',\n",
       " '__copy__',\n",
       " '__deepcopy__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__iadd__',\n",
       " '__imul__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__isub__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__mul__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__radd__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__sub__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_annotation',\n",
       " '_associate_gene',\n",
       " '_check_bounds',\n",
       " '_dissociate_gene',\n",
       " '_genes',\n",
       " '_gpr',\n",
       " '_id',\n",
       " '_lower_bound',\n",
       " '_metabolites',\n",
       " '_model',\n",
       " '_repr_html_',\n",
       " '_set_id_with_model',\n",
       " '_update_awareness',\n",
       " '_upper_bound',\n",
       " 'add_metabolites',\n",
       " 'annotation',\n",
       " 'boundary',\n",
       " 'bounds',\n",
       " 'build_reaction_from_string',\n",
       " 'build_reaction_string',\n",
       " 'check_mass_balance',\n",
       " 'compartments',\n",
       " 'copy',\n",
       " 'delete',\n",
       " 'flux',\n",
       " 'flux_expression',\n",
       " 'forward_variable',\n",
       " 'functional',\n",
       " 'gene_name_reaction_rule',\n",
       " 'gene_reaction_rule',\n",
       " 'genes',\n",
       " 'get_coefficient',\n",
       " 'get_coefficients',\n",
       " 'get_compartments',\n",
       " 'gpr',\n",
       " 'id',\n",
       " 'knock_out',\n",
       " 'lower_bound',\n",
       " 'metabolites',\n",
       " 'model',\n",
       " 'name',\n",
       " 'notes',\n",
       " 'objective_coefficient',\n",
       " 'products',\n",
       " 'reactants',\n",
       " 'reaction',\n",
       " 'reduced_cost',\n",
       " 'remove_from_model',\n",
       " 'reverse_id',\n",
       " 'reverse_variable',\n",
       " 'reversibility',\n",
       " 'subsystem',\n",
       " 'subtract_metabolites',\n",
       " 'summary',\n",
       " 'update_genes_from_gpr',\n",
       " 'update_variable_bounds',\n",
       " 'upper_bound',\n",
       " 'x',\n",
       " 'y']"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(mod_sbml.reactions[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fce9a41f-89fa-4a2f-9638-fad78220189a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>ihuman</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1c145224cd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>4504</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>6064</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>2032</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*MAR13082 - 1.0*MAR13082_reverse_11d67</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytosol, Endoplasmic reticulum, Extracellular, Mitochondria, Peroxisome, Golgi apparatus, Lysosome, Nucleus, Inner mitochondria</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model ihuman at 0x1c145224cd0>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_RPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6a3cc6c9-012f-4e0d-be46-eb84a2423a74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>ihuman</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1c140c2c7f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>3943</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>5149</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>2050</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*MAR13082 - 1.0*MAR13082_reverse_11d67</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytosol, Endoplasmic reticulum, Extracellular, Mitochondria, Peroxisome, Golgi apparatus, Lysosome, Nucleus, Inner mitochondria</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model ihuman at 0x1c140c2c7f0>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_PRrod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9e460b48-dc2e-435a-a406-7930b900eff2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ethanol:NAD+ oxidoreductase'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_PRrod.reactions[0].name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "05513127-3e4d-4511-bc95-8ba243333d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "mod = load_matlab_model('C:/Users/prins/git/Human-GEM/model/Human-GEM.mat') # load HUMAN1 model (this takes really long!!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2675a22d-0858-4ef4-ab1c-1cb85883a9d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Glycolysis / Gluconeogenesis'"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_sbml.reactions[0].subsystem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1492930f-5bc3-4c9b-9f50-88397654e52b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>MAR03905</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>ethanol:NAD+ oxidoreductase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x1c14114dbd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>MAM01796c + MAM02552c --> MAM01249c + MAM02039c + MAM02553c</p>\n",
       "                    <p style='text-align:right'>ethanol + NAD+ --> acetaldehyde + H+ + NADH</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>ENSG00000147576 or ENSG00000172955 or ENSG00000180011 or ENSG00000187758 or ENSG00000196344 or...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction MAR03905 at 0x1c14114dbd0>"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_PRrod.reactions[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "dd6b5161-8dca-4cb8-a19a-99056fb74d04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mod_PRrod.reactions[1].subsystem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7560631b-d100-4b09-9aa9-8254f33e5f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "for rxn in mod.reactions:\n",
    "    if len(rxn.subsystem)>0:\n",
    "        print(rxn.subsystem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e9ba79de-79f7-4c36-ba5b-b6a6cc795777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__add__',\n",
       " '__class__',\n",
       " '__copy__',\n",
       " '__deepcopy__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__iadd__',\n",
       " '__imul__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__isub__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__mul__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__radd__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__sub__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_annotation',\n",
       " '_associate_gene',\n",
       " '_check_bounds',\n",
       " '_dissociate_gene',\n",
       " '_genes',\n",
       " '_gpr',\n",
       " '_id',\n",
       " '_lower_bound',\n",
       " '_metabolites',\n",
       " '_model',\n",
       " '_repr_html_',\n",
       " '_set_id_with_model',\n",
       " '_update_awareness',\n",
       " '_upper_bound',\n",
       " 'add_metabolites',\n",
       " 'annotation',\n",
       " 'boundary',\n",
       " 'bounds',\n",
       " 'build_reaction_from_string',\n",
       " 'build_reaction_string',\n",
       " 'check_mass_balance',\n",
       " 'compartments',\n",
       " 'copy',\n",
       " 'delete',\n",
       " 'flux',\n",
       " 'flux_expression',\n",
       " 'forward_variable',\n",
       " 'functional',\n",
       " 'gene_name_reaction_rule',\n",
       " 'gene_reaction_rule',\n",
       " 'genes',\n",
       " 'get_coefficient',\n",
       " 'get_coefficients',\n",
       " 'get_compartments',\n",
       " 'gpr',\n",
       " 'id',\n",
       " 'knock_out',\n",
       " 'lower_bound',\n",
       " 'metabolites',\n",
       " 'model',\n",
       " 'name',\n",
       " 'notes',\n",
       " 'objective_coefficient',\n",
       " 'products',\n",
       " 'reactants',\n",
       " 'reaction',\n",
       " 'reduced_cost',\n",
       " 'remove_from_model',\n",
       " 'reverse_id',\n",
       " 'reverse_variable',\n",
       " 'reversibility',\n",
       " 'subsystem',\n",
       " 'subtract_metabolites',\n",
       " 'summary',\n",
       " 'update_genes_from_gpr',\n",
       " 'update_variable_bounds',\n",
       " 'upper_bound',\n",
       " 'x',\n",
       " 'y']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(mod_PRrod.reactions[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "fc606dd2-da21-4df5-8b21-40b9fb246444",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "142"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([rxn.name for rxn in mod_RPE.reactions if 'ATP' in rxn.name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0e4e968e-04fa-47ea-b52f-fc6fc6efa3fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len([rxn.name for rxn in mod_PRrod.reactions if 'ATP' in rxn.name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "89f5979f-98b5-4a73-acec-7a23db70c40e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATP:acylglycerol 3-phosphotransferase', \"ATP:dephospho-CoA 3'-phosphotransferase\", 'ATP:1D-myo-inositol-1,3,4-trisphosphate 6-phosphotransferase', 'ATP:1D-myo-inositol-3,4,5,6-tetrakisphosphate 1-phosphotransferase', 'ATP:D-ribose-5-phosphate diphosphotransferase', 'ATP:N-acetyl-D-glucosamine 6-phosphotransferase', \"ATP:pantetheine-4'-phosphate adenylyltransferase\", 'ATP:1-phosphatidyl-1D-myo-inositol-5-phosphate 4-phosphotransferase', 'ATP:6-deoxy-L-galactose 1-phosphotransferase', 'ATP:1-phosphatidyl-1D-myo-inositol 3-phosphotransferase', 'Transport of 11, 12-EET, FATP', 'ATP:choline phosphotransferase', \"ATP:cytidine 5'-phosphotransferase\", 'ATP:1D-myo-inositol-1,3,4-trisphosphate 5-phosphotransferase', 'ATP:1D-myo-inositol-1,3,4,6-tetrakisphosphate 5-phosphotransferase']\n",
      "['ATP:1,2-diacylglycerol 3-phosphotransferase', \"ATP:adenylylsulfate 3'-phosphotransferase\", 'ATP:1D-myo-inositol-5-diphosphate-pentakisphosphate phosphotransferase', \"ATP:deoxycitidine 5'-phosphotransferase\", 'ATP:1D-myo-inositol 1,3,4,5,6-pentakisphosphate 2-phosphotransferase', 'ATP:D-xylulose 5-phosphotransferase', 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.203.3', 'ATP:beta-D-ribosylnicotinate 5-phosphotransferase']\n"
     ]
    }
   ],
   "source": [
    "ATP_RPE = [rxn.name for rxn in mod_RPE.reactions if 'ATP' in rxn.name]\n",
    "ATP_PR = [rxn.name for rxn in mod_PRrod.reactions if 'ATP' in rxn.name]\n",
    "ATP1 = list(set(ATP_RPE) - set(ATP_PR))\n",
    "print(ATP1)\n",
    "ATP2 = list(set(ATP_PR) - set(ATP_RPE))\n",
    "print(ATP2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "52cd00cf-b90c-44ad-9a35-f5307be52276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP:acylglycerol 3-phosphotransferase',\n",
       " \"ATP:dephospho-CoA 3'-phosphotransferase\",\n",
       " 'ATP:1D-myo-inositol-1,3,4-trisphosphate 6-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-3,4,5,6-tetrakisphosphate 1-phosphotransferase',\n",
       " 'ATP:D-ribose-5-phosphate diphosphotransferase',\n",
       " 'ATP:N-acetyl-D-glucosamine 6-phosphotransferase',\n",
       " \"ATP:pantetheine-4'-phosphate adenylyltransferase\",\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol-5-phosphate 4-phosphotransferase',\n",
       " 'ATP:6-deoxy-L-galactose 1-phosphotransferase',\n",
       " 'ATP:1-phosphatidyl-1D-myo-inositol 3-phosphotransferase',\n",
       " 'Transport of 11, 12-EET, FATP',\n",
       " 'ATP:choline phosphotransferase',\n",
       " \"ATP:cytidine 5'-phosphotransferase\",\n",
       " 'ATP:1D-myo-inositol-1,3,4-trisphosphate 5-phosphotransferase',\n",
       " 'ATP:1D-myo-inositol-1,3,4,6-tetrakisphosphate 5-phosphotransferase']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2aeaed52-6e05-47ab-91ab-acc1cdc8d41f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP:1,2-diacylglycerol 3-phosphotransferase',\n",
       " \"ATP:adenylylsulfate 3'-phosphotransferase\",\n",
       " 'ATP:1D-myo-inositol-5-diphosphate-pentakisphosphate phosphotransferase',\n",
       " \"ATP:deoxycitidine 5'-phosphotransferase\",\n",
       " 'ATP:1D-myo-inositol 1,3,4,5,6-pentakisphosphate 2-phosphotransferase',\n",
       " 'ATP:D-xylulose 5-phosphotransferase',\n",
       " 'ATP-Binding Cassette (ABC) Tcdb:3.A.1.203.3',\n",
       " 'ATP:beta-D-ribosylnicotinate 5-phosphotransferase']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATP2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "9b4c55fc-b582-400d-8d27-cc6261befc30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>MAM00003c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>(10Z)-heptadecenoic acid</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x1c1452251b0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C17H31O2</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    MAR10033, MAR00240, MAR00241\n",
       "                    </td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite MAM00003c at 0x1c1452251b0>"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mod_RPE.metabolites[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f5275bc9-ac6e-42bf-ab0c-4dd11d1cd178",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATP_met_RPE = [met.name for met in mod_RPE.metabolites if 'ATP' in met.name]\n",
    "ATP_met_PR = [met.name for met in mod_PRrod.metabolites if 'ATP' in met.name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2da020c1-6dbe-49d1-97f9-200ce2f8877e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'dATP',\n",
       " 'dATP',\n",
       " 'dATP']"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATP_met_RPE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "4332c05c-edb1-4120-a8b6-285baaa0e7ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'ATP',\n",
       " 'dATP',\n",
       " 'dATP',\n",
       " 'dATP']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATP_met_PR"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
